<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Collection of Tools to Analyze AVID-seq Data • djvdj</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Collection of Tools to Analyze AVID-seq Data">
<meta property="og:description" content="What the package does (one paragraph).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">djvdj</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="djvdj-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#djvdj-" class="anchor"></a>djvdj <img src="reference/figures/djvdj-logo.png" align="right" height="145">
</h1></div>
<!-- badges: start -->

<p>The goal of djvdj is to provide tools to analyze AVID-seq signals alongside single-cell VDJ sequencing data.</p>
<p><br></p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the development version of djvdj from <a href="https://github.com/rnabioco/djvdj">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"rnabioco/djvdj"</span><span class="op">)</span></pre></div>
<p><br></p>
</div>
<div id="vignette" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette" class="anchor"></a>Vignette</h2>
<p>Splenocytes from MD4 transgenic mice which have monoclonal B cells that all bind hen egg lysozyme (HEL) antigen were mixed with splenocytes from C57BL/6 mice at a 1:20 ratio. The cells were stained with the HEL AVID-tag and sequencing libraries were prepared to capture gene expression, B cell receptor sequences, and AVID-tag signals using the 10x Genomics 5’ immune profiling kit.</p>
<p><img src="reference/figures/README-rna_umap-1.png" width="100%"></p>
<p><br></p>
<div id="import-vdj-data" class="section level3">
<h3 class="hasAnchor">
<a href="#import-vdj-data" class="anchor"></a>Import VDJ data</h3>
<p><code>import_vdj</code> takes the output files from <code>cellranger vdj</code> and adds clonotype information to the meta.data for an existing Seurat object. For cells that do not have any VDJ sequencing data, NAs will be included in the meta.data. The <code>filter_contigs</code> argument will only include chains that have at least one contig that is full length and productive.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>
  sobj_in        <span class="op">=</span> <span class="va">so</span>,              <span class="co"># Seurat object                         </span>
  vdj_dir        <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">vdj_dir</span>,  <span class="co"># Directory containing cellranger output files</span>
  prefix         <span class="op">=</span> <span class="st">""</span>,              <span class="co"># Prefix to add to new meta.data columns</span>
  cell_prefix    <span class="op">=</span> <span class="st">""</span>,              <span class="co"># Prefix to add to cell barcodes</span>
  filter_contigs <span class="op">=</span> <span class="cn">TRUE</span>             <span class="co"># Only include chains with at least one productive contig</span>
<span class="op">)</span>

<span class="va">vdj_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chains"</span>, <span class="st">"cdr3"</span>, <span class="st">"n_chains"</span><span class="op">)</span>

<span class="va">so_vdj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="va">orig.ident</span>, <span class="va">nCount_RNA</span>, <span class="va">nFeature_RNA</span>, <span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 7,137 x 6</span>
<span class="co">#&gt;    orig.ident nCount_RNA nFeature_RNA chains  cdr3                      n_chains</span>
<span class="co">#&gt;    &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                        &lt;int&gt;</span>
<span class="co">#&gt;  1 AVID-seq          884          551 IGH;IGK CVKGYDYDWYFDVW;CLQYDNLWTF        2</span>
<span class="co">#&gt;  2 AVID-seq         3061          970 &lt;NA&gt;    &lt;NA&gt;                            NA</span>
<span class="co">#&gt;  3 AVID-seq         1297          677 IGH;IGK CARGRLGYAMDYW;CQHFWSTPWTF        2</span>
<span class="co">#&gt;  4 AVID-seq         1570          848 &lt;NA&gt;    &lt;NA&gt;                            NA</span>
<span class="co">#&gt;  5 AVID-seq         2277          818 &lt;NA&gt;    &lt;NA&gt;                            NA</span>
<span class="co">#&gt;  6 AVID-seq         1320          566 &lt;NA&gt;    &lt;NA&gt;                            NA</span>
<span class="co">#&gt;  7 AVID-seq          570          348 &lt;NA&gt;    &lt;NA&gt;                            NA</span>
<span class="co">#&gt;  8 AVID-seq          909          489 IGH;IGK CTVSYTKDWYFDVW;CAQNLELPL…        2</span>
<span class="co">#&gt;  9 AVID-seq         1072          588 IGH;IGK CARSYDYDPLYYAMDYW;CLQSDN…        2</span>
<span class="co">#&gt; 10 AVID-seq         1143          594 IGH;IGK CARSRLAYW;CLQYASSPFTF            2</span>
<span class="co">#&gt; # … with 7,127 more rows</span></pre></div>
<p><br></p>
</div>
<div id="filtering" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering" class="anchor"></a>Filtering</h3>
<p><code>filter_vdj</code> allows you to filter a Seurat object using the added clonotype information or any other columns present in the meta.data. For cells with multiple chains, the information for each chain is stored as a single row, separated by a “;”. When filtering, columns with VDJ data will be expanded based on the delimiter “;”. The columns that are expanded for filtering can be specified with the <code>split_cols</code> argument. By default filtering is only performed on cells that include VDJ data.</p>
<p>Filter to only include cells with paired light and heavy chains.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">so_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in <span class="op">=</span> <span class="va">so_vdj</span>,                                              <span class="co"># Seurat object</span>
  filt    <span class="op">=</span> <span class="st">"IGH"</span> <span class="op">%in%</span> <span class="va">chains</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"IGK"</span>, <span class="st">"IGL"</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">chains</span><span class="op">)</span>  <span class="co"># Expression for filtering</span>
<span class="op">)</span>

<span class="va">so_filt</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3,353 x 3</span>
<span class="co">#&gt;    chains      cdr3                                  n_chains</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;                                    &lt;int&gt;</span>
<span class="co">#&gt;  1 IGH;IGK     CVKGYDYDWYFDVW;CLQYDNLWTF                    2</span>
<span class="co">#&gt;  2 IGH;IGK     CARGRLGYAMDYW;CQHFWSTPWTF                    2</span>
<span class="co">#&gt;  3 IGH;IGK     CTVSYTKDWYFDVW;CAQNLELPLTF                   2</span>
<span class="co">#&gt;  4 IGH;IGK     CARSYDYDPLYYAMDYW;CLQSDNLPLTF                2</span>
<span class="co">#&gt;  5 IGH;IGK     CARSRLAYW;CLQYASSPFTF                        2</span>
<span class="co">#&gt;  6 IGH;IGK;IGL CAKRGYSNSLDYW;CQHFWSTPYTF;CALWYSNHWVF        3</span>
<span class="co">#&gt;  7 IGH;IGK     CANPITTAEGWYFDVW;CLQHGESPYTF                 2</span>
<span class="co">#&gt;  8 IGH;IGK     CARSYGYAMDYW;CWQGTHFPYTF                     2</span>
<span class="co">#&gt;  9 IGH;IGK     CARWVYGSAWFAYW;CMQHLEYPFTF                   2</span>
<span class="co">#&gt; 10 IGH;IGK     CARSHGYDFYAMDYW;CQHFWGTPRTF                  2</span>
<span class="co">#&gt; # … with 3,343 more rows</span></pre></div>
<p><br></p>
<p>Instead of filtering, <code>filter_vdj</code> can also add new cell labels to the object meta.data using the <code>new_col</code> argument. Here a new column is added indicating whether each cell has a paired heavy and light chain. This is useful for plotting.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in <span class="op">=</span> <span class="va">so_vdj</span>,                                               <span class="co"># Seurat object</span>
  filt    <span class="op">=</span> <span class="st">"IGH"</span> <span class="op">%in%</span> <span class="va">chains</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"IGK"</span>, <span class="st">"IGL"</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">chains</span><span class="op">)</span>,  <span class="co"># Condition to use for filtering</span>
  new_col <span class="op">=</span> <span class="st">"Paired"</span>,                                             <span class="co"># Name of new column</span>
  true    <span class="op">=</span> <span class="st">"paired"</span>,                                             <span class="co"># Value when condition is TRUE</span>
  false   <span class="op">=</span> <span class="st">"unpaired"</span>                                            <span class="co"># Value when condition is FALSE</span>
<span class="op">)</span>

<span class="va">vdj_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">vdj_cols</span>, <span class="st">"Paired"</span><span class="op">)</span>

<span class="va">so_vdj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3,820 x 4</span>
<span class="co">#&gt;    chains      cdr3                                  n_chains Paired  </span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;                                    &lt;int&gt; &lt;chr&gt;   </span>
<span class="co">#&gt;  1 IGH;IGK     CVKGYDYDWYFDVW;CLQYDNLWTF                    2 paired  </span>
<span class="co">#&gt;  2 IGH;IGK     CARGRLGYAMDYW;CQHFWSTPWTF                    2 paired  </span>
<span class="co">#&gt;  3 IGH;IGK     CTVSYTKDWYFDVW;CAQNLELPLTF                   2 paired  </span>
<span class="co">#&gt;  4 IGH;IGK     CARSYDYDPLYYAMDYW;CLQSDNLPLTF                2 paired  </span>
<span class="co">#&gt;  5 IGH;IGK     CARSRLAYW;CLQYASSPFTF                        2 paired  </span>
<span class="co">#&gt;  6 IGH;IGK;IGL CAKRGYSNSLDYW;CQHFWSTPYTF;CALWYSNHWVF        3 paired  </span>
<span class="co">#&gt;  7 IGK         CQQWSSNPLTF                                  1 unpaired</span>
<span class="co">#&gt;  8 IGH;IGK     CANPITTAEGWYFDVW;CLQHGESPYTF                 2 paired  </span>
<span class="co">#&gt;  9 IGH;IGK     CARSYGYAMDYW;CWQGTHFPYTF                     2 paired  </span>
<span class="co">#&gt; 10 IGH;IGK     CARWVYGSAWFAYW;CMQHLEYPFTF                   2 paired  </span>
<span class="co">#&gt; # … with 3,810 more rows</span></pre></div>
<p><img src="reference/figures/README-pair_umap-1.png" width="100%"></p>
<p><br></p>
<p>More complicated statements referring to meta.data columns can be used for the <code>filt</code>, <code>true</code>, and <code>false</code> arguments. For more detailed analysis of the chains detected for each cell, a new cell label can be created for only the unique chains.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in <span class="op">=</span> <span class="va">so_vdj</span>,                               <span class="co"># Seurat object</span>
  new_col <span class="op">=</span> <span class="st">"uniq_chains"</span>,                        <span class="co"># Name of new column</span>
  true    <span class="op">=</span> <span class="fu">str_c</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chains</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>  <span class="co"># Value when condition is TRUE</span>
<span class="op">)</span>

<span class="va">vdj_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">vdj_cols</span>, <span class="st">"uniq_chains"</span><span class="op">)</span>

<span class="va">so_vdj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span>, <span class="va">n_chains</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 526 x 5</span>
<span class="co">#&gt;    chains       cdr3                                 n_chains Paired uniq_chains</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;chr&gt;                                   &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;      </span>
<span class="co">#&gt;  1 IGH;IGK;IGL  CAKRGYSNSLDYW;CQHFWSTPYTF;CALWYSNHW…        3 paired IGH_IGK_IGL</span>
<span class="co">#&gt;  2 IGH;IGH;IGK… CARGDYW;CTTWLRLRSFAYW;CHQRSSYPCTF;C…        4 paired IGH_IGK    </span>
<span class="co">#&gt;  3 IGH;IGK;IGK  CAKPRYYYGSSFYAMDYW;CQQGNTLPFTF;CAQN…        3 paired IGH_IGK    </span>
<span class="co">#&gt;  4 IGH;IGK;IGK  CARGPYYTNGGAMDYW;CQHFWSTPFTF;CLQYDE…        3 paired IGH_IGK    </span>
<span class="co">#&gt;  5 IGH;IGH;IGK… CARSYPYFDYW;CARSSITTVSDYW;CQQHNEYPY…        4 paired IGH_IGK    </span>
<span class="co">#&gt;  6 IGH;IGK;IGK  CALDSSGFAYW;CQQYWSTPTF;CHQYHRSPLTF          3 paired IGH_IGK    </span>
<span class="co">#&gt;  7 IGH;IGH;IGK… CARHDGLPGAMDYW;CARPLTIAGSAMDYW;CQQG…        4 paired IGH_IGK    </span>
<span class="co">#&gt;  8 IGH;IGH;IGK… CAEGSSNWYFDVW;CARDGSSPFDYW;CQQHNEYP…        4 paired IGH_IGK    </span>
<span class="co">#&gt;  9 IGH;IGK;IGK  CTSPPYEGYYAMDYW;CFQGSHVPPTF;CKQSYNL…        3 paired IGH_IGK    </span>
<span class="co">#&gt; 10 IGH;IGH;IGK… CTRLLTGYYFDYW;CARRYYYGSSWNYFDYW;CQQ…        4 paired IGH_IGK    </span>
<span class="co">#&gt; # … with 516 more rows</span></pre></div>
<p><img src="reference/figures/README-chains_umap-1.png" width="100%"></p>
<p><br></p>
<p>MD4 B cells are expected to have IGK chains with the CDR3 amino acid sequence CQQSNSWPYTF. Using <code>filter_vdj</code> we can visualize which cells have this sequence.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># Add new cell labels to meta.data</span>
<span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in <span class="op">=</span> <span class="va">so_vdj</span>,                                   <span class="co"># Seurat object</span>
  filt    <span class="op">=</span> <span class="st">"CQQSNSWPYTF"</span> <span class="op">%in%</span> <span class="va">cdr3</span><span class="op">[</span><span class="va">chains</span> <span class="op">==</span> <span class="st">"IGK"</span><span class="op">]</span>,  <span class="co"># Condition to use for filtering</span>
  new_col <span class="op">=</span> <span class="st">"IGK_seq"</span>,                                <span class="co"># Name of new column</span>
  true    <span class="op">=</span> <span class="st">"CQQSNSWPYTF"</span>,                            <span class="co"># Value when condition is TRUE</span>
  false   <span class="op">=</span> <span class="st">"other"</span>                                   <span class="co"># Value when condition is FALSE</span>
<span class="op">)</span>

<span class="va">vdj_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">vdj_cols</span>, <span class="st">"IGK_seq"</span><span class="op">)</span>

<span class="va">so_vdj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3,820 x 6</span>
<span class="co">#&gt;    chains     cdr3                          n_chains Paired  uniq_chains IGK_seq</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;                            &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;  </span>
<span class="co">#&gt;  1 IGH;IGK    CVKGYDYDWYFDVW;CLQYDNLWTF            2 paired  IGH_IGK     other  </span>
<span class="co">#&gt;  2 IGH;IGK    CARGRLGYAMDYW;CQHFWSTPWTF            2 paired  IGH_IGK     other  </span>
<span class="co">#&gt;  3 IGH;IGK    CTVSYTKDWYFDVW;CAQNLELPLTF           2 paired  IGH_IGK     other  </span>
<span class="co">#&gt;  4 IGH;IGK    CARSYDYDPLYYAMDYW;CLQSDNLPLTF        2 paired  IGH_IGK     other  </span>
<span class="co">#&gt;  5 IGH;IGK    CARSRLAYW;CLQYASSPFTF                2 paired  IGH_IGK     other  </span>
<span class="co">#&gt;  6 IGH;IGK;I… CAKRGYSNSLDYW;CQHFWSTPYTF;CA…        3 paired  IGH_IGK_IGL other  </span>
<span class="co">#&gt;  7 IGK        CQQWSSNPLTF                          1 unpair… IGK         other  </span>
<span class="co">#&gt;  8 IGH;IGK    CANPITTAEGWYFDVW;CLQHGESPYTF         2 paired  IGH_IGK     other  </span>
<span class="co">#&gt;  9 IGH;IGK    CARSYGYAMDYW;CWQGTHFPYTF             2 paired  IGH_IGK     other  </span>
<span class="co">#&gt; 10 IGH;IGK    CARWVYGSAWFAYW;CMQHLEYPFTF           2 paired  IGH_IGK     other  </span>
<span class="co">#&gt; # … with 3,810 more rows</span></pre></div>
<p><img src="reference/figures/README-seq_umap-1.png" width="100%"></p>
<p><br></p>
</div>
<div id="repertoire-stats" class="section level3">
<h3 class="hasAnchor">
<a href="#repertoire-stats" class="anchor"></a>Repertoire stats</h3>
<p>The functions <code>calc_diversity</code> and <code>calc_jaccard</code> will calculate repertoire diversity and repertoire overlap on a per-cluster basis. These functions can be given any meta.data column containing cell labels to use for calculations.</p>
<p>Calculate repertoire diversity with <code>calc_diversity</code>. The inverse Simpson index is used to measure diversity for each cluster.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/calc_diversity.html">calc_diversity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_vdj</span>,          <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,  <span class="co"># meta.data column containing clonotype ids</span>
  cluster_col   <span class="op">=</span> <span class="st">"type_mouse"</span>,    <span class="co"># meta.data column containing cell labels</span>
  prefix        <span class="op">=</span> <span class="st">""</span>               <span class="co"># Prefix to add to new meta.data columns</span>
<span class="op">)</span></pre></div>
<p><img src="reference/figures/README-div_umap-1.png" width="100%"></p>
<p><br></p>
<p>Calculate repertoire overlap with <code>calc_jaccard</code> for the cell groups present in the <code>cluster_col</code>. Using the <code>return_seurat</code> argument, <code>calc_jaccard</code> can also output a matrix for plotting.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/calc_overlap.html">calc_overlap</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_vdj</span>,             <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,     <span class="co"># meta.data column containing clonotype ids</span>
  cluster_col   <span class="op">=</span> <span class="st">"seurat_clusters"</span>,  <span class="co"># meta.data column containing cell labels</span>
  prefix        <span class="op">=</span> <span class="st">"x"</span>,                <span class="co"># Prefix to add to new meta.data columns </span>
  return_seurat <span class="op">=</span> <span class="cn">TRUE</span>                <span class="co"># Return Seurat object with results added to meta.data</span>
<span class="op">)</span></pre></div>
<p><img src="reference/figures/README-jaccard_umap-1.png" width="100%"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>First Last <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/YOUR-ORCID-ID" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/rnabioco/djvdj/actions"><img src="https://github.com/rnabioco/djvdj/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by First Last.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
