---
output: github_document
editor_options: 
  chunk_output_type: inline
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse  = TRUE,
  comment   = "#>",
  fig.path  = "man/figures/README-",
  warning   = FALSE,
  message   = FALSE
)
```

# djvdj <img src="man/figures/djvdj-logo.png" align="right" height="145">

<!-- badges: start -->
[![R build status](https://github.com/rnabioco/djvdj/workflows/R-CMD-check/badge.svg)](https://github.com/rnabioco/djvdj/actions)
<!-- badges: end -->

The goal of djvdj is to provide tools to analyze AVID-seq signals alongside single-cell VDJ sequencing data.

<br>

## Installation

You can install the development version of djvdj from [GitHub](https://github.com/rnabioco/djvdj) with:

```{r, eval = FALSE}
# install.packages("devtools")
devtools::install_github("rnabioco/djvdj")
```

```{r "Setup", include = FALSE}
# Packages
library(tidyverse)
library(Seurat)
library(djvdj)
library(cowplot)
library(colorblindr)
library(here)

# Themes
umap_theme <- vdj_theme() +
  theme(
    plot.title      = element_text(size = 12),
    legend.position = "top",
    legend.title    = element_blank(),
    axis.title      = element_blank(),
    axis.line       = element_blank(),
    axis.ticks      = element_blank(),
    axis.text       = element_blank()
  )

# Legend guides
create_guide <- function(size = 3.5, shape = 16, nrow = NULL,
                         ncol = NULL, ...) {
  guide_legend(
    nrow = nrow,
    ncol = ncol,
    override.aes = list(
      size   = size,
      shape  = shape,
      ...
    )
  )
}

col_guide <- create_guide(shape = 16, stroke = NULL, color = NULL)

# Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[1:4], "#d7301f", 
  palette_OkabeIto[5:6], "#6A51A3", 
  palette_OkabeIto[7:8], "#875C04"
)

# Set default color palette
create_gradient <- function(cols_in, n = NULL) {
  if (is.null(n)) {
    n <- length(cols_in)
  }
  
  colorRampPalette(cols_in)(n)
}

create_col_fun <- function(cols_in) {
  function(n = NULL) {
    create_gradient(cols_in, n)
  }
}

get_cols <- create_col_fun(ito_cols)
```

<br>

## Vignette

Write a description here...

```{r "rna_umap", fig.width = 12.5, fig.height = 5, echo = FALSE}
load(here("data/so_tcr.rda"))

key_umap <- so_tcr %>%
  plot_features(
    feature     = "orig.ident",
    pt_size     = 0.02,
    plot_colors = get_cols()
  ) +
  ggtitle("Cell type") +
  guides(color = create_guide()) +
  umap_theme +
  theme(
    legend.position = "right",
    legend.title    = element_blank()
  )

clust_umap <- so_tcr %>%
  plot_features(
    feature     = "seurat_clusters",
    pt_size     = 0.02,
    plot_colors = get_cols(22)
  ) +
  ggtitle("Cluster") +
  guides(color = create_guide()) +
  umap_theme +
  theme(
    legend.position = "right",
    legend.title    = element_blank()
  )

plot_grid(
  key_umap, clust_umap,
  rel_widths = c(1, 0.9),
  align      = "h",
  axis       = "tb"
)
```

<br>

### Import VDJ data

`import_vdj` takes the output files from `cellranger vdj` and adds clonotype information to the meta.data for an existing Seurat object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a ";" (or a character specified by `sep`). For cells that do not have any VDJ sequencing data, NAs will be added to the meta.data.

If the Seurat object contains data for multiple runs, a vector containing paths to the VDJ data for each sample can be given. If multiple paths are provided, cell prefixes should be included as names for the vector. 

```{r "import_vdj"}
# Create vector of paths for cellranger output
samples <- levels(so_tcr$orig.ident)
paths   <- file.path("data", str_c(samples, "_TCR"))

names(paths) <- str_c(samples, "_GE")

# Import VDJ data
so_tcr <- import_vdj(
  sobj_in        = so_tcr,  # Seurat object
  vdj_dir        = paths,   # Directories containing cellranger output files
  prefix         = "",      # Prefix to add to new meta.data columns
  filter_contigs = TRUE     # Only include chains with at least one productive contig
)

# Take a look at the meta.data
vdj_cols <- c(
  "clonotype_id", "cdr3",
  "chains", "v_gene", 
  "j_gene", "reads",
  "umis"
)

so_tcr@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

### Quality Control

#### Paired Chains

The V(D)J data imported from Cell Ranger will include clonotypes that do not have paired alpha and beta chains. Using the `mutate_vdj` function, we can add a new cell label to the meta.data to allow for easy visualization of these cells.

In this example we create a new column that summarizes the unique chains identified for each cell. We can then overlay these cell labels on a UMAP projection.

```{r}
so_tcr <- mutate_vdj(
  sobj_in     = so_tcr,                                # Seurat object
  uniq_chains = str_c(unique(chains), collapse = "_")  # New column
)

# Take a look at the meta.data
so_tcr@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id), n_chains > 2) %>%
  select(all_of(vdj_cols), uniq_chains)
```

```{r "chains_umap", fig.width = 6.5, fig.height = 4.5, echo = FALSE}
# Plot colors
chains_cols <- so_tcr@meta.data %>%
  pull(uniq_chains) %>%
  unique() %>%
  na.omit()

chains_cols <- set_names(
  x  = get_cols()[seq_along(chains_cols)],
  nm = chains_cols
)

# Chains UMAP
chains_umap <- so_tcr %>%
  plot_features(
    feature     = "uniq_chains",
    pt_size     = 0.02,
    plot_colors = chains_cols
  ) +
  ggtitle("Unique chains") +
  guides(color = create_guide(nrow = 2)) +
  umap_theme

# Chains bar graphs
chains_bars <- so_tcr %>%
  plot_cell_count(
    x           = "orig.ident",
    fill_col    = "uniq_chains",
    plot_colors = chains_cols
  ) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  theme(
    plot.margin     = unit(c(120, 10, 10, 10), "pt"),
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1)
  )

# Final figure
plot_grid(
  chains_umap, chains_bars,
  rel_widths = c(1, 0.55),
  nrow       = 1
)
```

<br>

Clonotypes that lack paired alpha and beta chains can be removed from the meta.data using `filter_vdj`. With this function cells that match the provided condition are filtered from the object. Setting `filter_cells` to `FALSE` will just remove the V(D)J data without completely removing these cells.

```{r}
so_filt <- filter_vdj(
  sobj_in      = so_tcr,                            # Seurat object
  filt         = all(c("TRA", "TRB") %in% chains),  # Condition for filtering
  filter_cells = FALSE                              # Should cells be removed from object
)

# Take a look at the meta.data
so_filt@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

#### Read Support

The read support for each chain can be visualized with the `plot_reads` function. This will create plots summarizing the number of UMIs and total reads that were obtained for each chain.

```{r "read_support", fig.width = 6.5, fig.height = 3}
plot_reads(
  sobj_in      = so_tcr,        # Seurat object
  chain_col    = "chains",      # Column containing chains for each cell
  cluster_col  = "orig.ident",  # Column containing labels to group by
  plot_colors  = ito_cols       # Plot colors
)
```

<br>

Clonotypes that do not have adequate UMI support can be filtered from the object using `filter_vdj`. In this example we filter for clonotypes that have chains supported by at least two UMIs.

```{r}
so_tcr <- filter_vdj(
  sobj_in      = so_tcr,         # Seurat object
  filt         = all(umis > 1),  # Condition for filtering
  filter_cells = FALSE           # Should cells be removed from object
)

# Take a look at the meta.data
so_tcr@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

### Clonotype Abundance

To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the `calc_abundance` function.

```{r}
so_tcr <- calc_abundance(
  sobj_in       = so_tcr,        # Seurat object
  clonotype_col = "cdr3",        # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident",  # meta.data column containing cell labels
  prefix        = ""             # Prefix to add to new meta.data columns
)
```

<br>

For each 'calculation' function provided by `djvdj`, there is a matching `plot` function that will generate a summary plot. The `calc_abundance` function will plot clonotypes ranked by abundance.

```{r "abund_plots", fig.width = 6, fig.height = 3}
plot_abundance(
  sobj_in       = so_tcr,        # Seurat object
  clonotype_col = "cdr3",        # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident",  # meta.data column containing cell labels
  
  plot_colors = ito_cols,        # Plot colors
  yaxis       = "percent",       # Units to plot
  label_col   = "cdr3",          # meta.data column containing labels
  n_labels    = 1,               # Number of top clonotypes to label
  size        = 1                # Additional ggplot options
) +
  theme(legend.title = element_blank())
```

<br>

### Repertoire Diversity

The function `calc_diversity` will calculate repertoire diversity on a per-cluster basis. Using the `cluster_col` argument, any meta.data column containing cell labels can be used for calculations. `calc_diversity` uses the R package `abdiv` for performing diversity calculations and any `abdiv` diversity function can be specified using the `method` argument.

Possible methods for calculating diversity include:

```{r, comment = "", echo = F}
abdiv::alpha_diversities
```
<br>

In this example we are calculating the Shannon diversity for each sample in the orig.ident meta.data column.

```{r}
so_tcr <- calc_diversity(
  sobj_in       = so_tcr,         # Seurat object
  clonotype_col = "cdr3",         # meta.data column containing clonotype ids
  cluster_col   = "orig.ident",   # meta.data column containing cell labels
  method        = abdiv::shannon  # abdiv method to use
)
```

<br>

The `plot_diversity` function will create plots summarizing repertoire diversity for each sample. A named list of functions to run can also be passed to plot multiple metrics. Four different diversity metrics are shown in the example below.

```{r "div_plots", fig.width = 7.5, fig.height = 3}
# Metrics to plot
fns <- list(
  "simpson"     = abdiv::simpson,
  "shannon"     = abdiv::shannon,
  "margalef"    = abdiv::margalef,
  "menhinick"   = abdiv::menhinick,
  "brillouin_d" = abdiv::brillouin_d
)

plot_diversity(
  sobj_in       = so_tcr,        # Seurat object
  clonotype_col = "cdr3",        # meta.data column containing clonotype ids
  cluster_col   = "orig.ident",  # meta.data column containing cell labels
  method        = fns,           # abdiv method to use
  plot_colors   = ito_cols
) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

<br>

### Repertoire Overlap

To compare repertoires for different samples or clusters, `calc_similarity` can calculate a variety of different similarity metrics. The `cluster_col` should be used to specify the meta.data column containing cell labels for comparison. Like `calc_diversity`, an `abdiv` function can be specified with the `method` argument.

Possible methods for calculating repertoire similarity include:

```{r, comment = "", echo = F}
abdiv::beta_diversities
```

<br>

By default `calc_similarity` will add a new meta.data column for each comparison. In this example we are calculating the jaccard dissimilarity index for all combinations of cell labels present in the `orig.ident` column.

```{r}
so_tcr <- calc_similarity(
  sobj_in       = so_tcr,          # Seurat object
  clonotype_col = "cdr3",          # meta.data column containing clonotype ids
  cluster_col   = "orig.ident",    # meta.data column containing cell labels
  method        = abdiv::jaccard,  # abdiv method to use
  prefix        = "jcrd_",         # Prefix to add to new meta.data columns 
  return_seurat = TRUE             # Return Seurat object with results added to meta.data
)
```

<br>

A heatmap summarizing the results can be generated using the `plot_similarity` function. Here we are creating two heatmaps, one to compare the different samples and one to compare cell clusters.

```{r "sim_plots", fig.width = 10, fig.height = 4}
heat_theme <- theme(
  legend.title = element_blank(),
  legend.text  = element_text(size = 8)
)

# Sample heatmap
ident_heat <- plot_similarity(
  sobj_in       = so_tcr,                 # Seurat object
  clonotype_col = "cdr3",                 # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident",           # meta.data column containing cell labels
  method        = abdiv::jaccard,         # Method to use
  plot_colors   = c("grey90", "#009E73")  # Plot colors
) +
  heat_theme

# Cluster heatmap
clust_heat <- plot_similarity(
  sobj_in       = so_tcr,
  clonotype_col = "cdr3",
  cluster_col   = "seurat_clusters",
  method        = abdiv::jaccard,
  plot_colors   = c("grey90", "#56B4E9"),  
  size          = 0.2,                    # Additional ggplot options
  color         = "white"                 # Additional ggplot options
) +
  heat_theme +
  theme(axis.text.x  = element_text(angle = 0))

# Combine heatmaps
plot_grid(ident_heat, clust_heat, align = "h")
```

<br>

### Gene Usage

The V(D)J data imported from Cell Ranger also includes the specific genes detected for each cell. The function `calc_usage` can be used to calculate the fraction of cells that express different V(D)J genes. This function will produce a table summarizing the results. To only include results for a certain chain, the `chain` and `chain_col` arguments can be used to specify the meta.data column containing the chains detected for each cell. By default the results for all chains will be included.

In this example we are summarizing the usage of different V genes for the TRB chain 

```{r}
calc_usage(
  sobj_in     = so_tcr,        # Seurat object
  gene_cols   = "v_gene",      # meta.data column containing genes
  cluster_col = "orig.ident",  # meta.data column containing cell labels
  chain       = "TRB",         # Chain to use for filtering genes
  chain_col   = "chains"       # meta.data column containing chains identified for each cell
)
```

<br>

The function `plot_usage` can be used to summarize these results. Using the `yaxis` argument, the percent or absolute count (frequency) can be used for plotting. The genes plotted can also be selected using the `plot_genes` argument, or the number of top genes (most frequent) to plot can be specified with `n_genes`.

```{r "usage_plots_1", fig.width = 11, fig.height = 3.5}
plot_usage(
  sobj_in     = so_tcr,                # Seurat object
  gene_cols   = "v_gene",              # meta.data column(s) containing genes
  cluster_col = "orig.ident",          # meta.data column containing cell labels
  type        = "bar",                 # Type of plot
  chain       = "TRB",                 # Chain to use for filtering genes
  chain_col   = "chains",              # meta.data column containing chains
  
  yaxis       = "percent",             # Units to plot
  plot_colors = ito_cols,              # Colors to use for heatmap
  plot_genes  = NULL,                  # A list of genes to plot
  n_genes     = NULL,                  # The number of top genes to plot
  
  size        = 0.2,                   # Additional ggplot options
  color       = "white"                # Additional ggplot options
)
```

<br>

By passing multiple columns to `gene_cols`, the frequency that different genes are used together can also be summarized.

```{r, eval = F}
calc_usage(
  sobj_in     = so_tcr,                 # Seurat object
  gene_cols   = c("v_gene", "j_gene"),  # meta.data column(s) containing genes
  cluster_col = "orig.ident",           # meta.data column containing cell labels
  chain       = "TRB",                  # Chain to use for filtering genes
  chain_col   = "chains"                # meta.data column containing chains
)
```

<br>

When multiple gene columns are passed to `plot_usage`, a list of plots will be returned, one for each cell label in the `cluster_col` column.

```{r "usage_plots_2", fig.width = 12, fig.height = 8}
ggs <- plot_usage(
  sobj_in     = so_tcr,                 # Seurat object
  gene_cols   = c("v_gene", "j_gene"),  # meta.data column(s) containing genes
  cluster_col = "orig.ident",           # meta.data column containing cell labels
  chain       = "TRB",                  # Chain to use for filtering genes
  chain_col   = "chains",               # meta.data column containing chains identified
  plot_colors = c("grey90", "#6A51A3")  # Colors to use for heatmap
) %>%
  imap(~ .x + ggtitle(.y))

plot_grid(plotlist = ggs)
```





```{r "enclone_test", eval = FALSE, echo = FALSE}
paths        <- dir("data", pattern = "[0-9]_AB")
names(paths) <- str_remove(paths, "_b_Cell$|_bCellEnriched$")
paths        <- map_chr(paths, ~ file.path("data", .x))




# Import enclone results
en_df <- read_csv(here("enclone_out.csv")) %>%
  select(all_of(en_cols)) %>%
  mutate(barcodes = str_split(barcodes, ",")) %>%
  unnest(barcodes) %>%
  rename(cell_id = barcodes) %>%
  filter(group_ncells > 1)

# Merge enclone groups with meta.data
en_df <- so_tcr@meta.data %>%
  as_tibble(rownames = "cell_id") %>%
  left_join(en_df, by = "cell_id") %>%
  mutate(
    group_id = as.character(group_id),
    group_id = replace_na(group_id, "NA")
  )

# Group colors
grp_cols <- en_df %>%
  pull(group_id) %>%
  unique()

grp_cols <- set_names(
  x  = c(ito_cols, darken(ito_cols, 0.2))[seq_along(grp_cols)],
  nm = grp_cols
)

grp_cols["NA"] <- "white"

# Overlay enclone groups on UMAP
en_df %>%
  mutate(group_id = fct_relevel(group_id, names(grp_cols))) %>%
  plot_features(
    x          = "vdjUMAP_1",
    y          = "vdjUMAP_2",
    feature    = "group_id",
    pt_outline = 0.3,
    plot_colors  = grp_cols
  ) +
  umap_theme +
  guides(color = guide_legend(override.aes = list(size = 3))
```





