---
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r "setup", include = FALSE}
# Chunk opts
knitr::opts_chunk$set(
  collapse  = TRUE,
  comment   = "#>",
  fig.path  = "man/figures/README-",
  warning   = FALSE,
  message   = FALSE
)

# Packages
library(tidyverse)
library(Seurat)
library(djvdj)
library(cowplot)
library(colorblindr)
library(here)

# Themes
create_guide <- function(size = 3.5, shape = 16, nrow = NULL,
                         ncol = NULL, ...) {
  guide_legend(
    nrow = nrow,
    ncol = ncol,
    override.aes = list(
      size   = size,
      shape  = shape,
      ...
    )
  )
}

umap_theme <- vdj_theme() +
  theme(
    plot.title      = element_text(size = 12),
    legend.position = "top",
    legend.title    = element_blank(),
    axis.title      = element_blank(),
    axis.line       = element_blank(),
    axis.ticks      = element_blank(),
    axis.text       = element_blank()
  )

# Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[1:4], "#d7301f", 
  palette_OkabeIto[5:6], "#6A51A3", 
  palette_OkabeIto[7:8], "#875C04"
)

# Set default color palette
create_gradient <- function(cols_in, n = NULL) {
  if (is.null(n)) {
    n <- length(cols_in)
  }
  
  colorRampPalette(cols_in)(n)
}

create_col_fun <- function(cols_in) {
  function(n = NULL) {
    create_gradient(cols_in, n)
  }
}

get_cols <- create_col_fun(ito_cols)
```

# djvdj <img src="man/figures/djvdj-logo.png" align="right" height="145">

<!-- badges: start -->
[![R build status](https://github.com/rnabioco/djvdj/workflows/R-CMD-check/badge.svg)](https://github.com/rnabioco/djvdj/actions)
<!-- badges: end -->

The goal of djvdj is to provide tools to analyze AVID-seq signals alongside single-cell VDJ sequencing data.

<br>

## Installation

You can install the development version of djvdj from [GitHub](https://github.com/rnabioco/djvdj) with:

```{r, eval = FALSE}
# install.packages("devtools")
devtools::install_github("rnabioco/djvdj")
```

<br>

## Vignette

Write a description here...

```{r "rna_umap", fig.width = 12.5, fig.height = 5, echo = FALSE}
load(here("data/so_tcr.rda"))

key_umap <- so_tcr %>%
  plot_features(
    feature     = "orig.ident",
    pt_size     = 0.02,
    plot_colors = get_cols()
  ) +
  ggtitle("Cell type") +
  guides(color = create_guide()) +
  umap_theme +
  theme(
    legend.position = "right",
    legend.title    = element_blank()
  )

clust_umap <- so_tcr %>%
  plot_features(
    feature     = "seurat_clusters",
    pt_size     = 0.02,
    plot_colors = get_cols(22)
  ) +
  ggtitle("Cluster") +
  guides(color = create_guide()) +
  umap_theme +
  theme(
    legend.position = "right",
    legend.title    = element_blank()
  )

plot_grid(
  key_umap, clust_umap,
  rel_widths = c(1, 0.9),
  align      = "h",
  axis       = "tb"
)
```

<br>

### Import VDJ data

`import_vdj` takes the output files from `cellranger vdj` and adds clonotype information to the meta.data for an existing Seurat object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a ";" (or a character specified by `sep`). For cells that do not have any VDJ sequencing data, NAs will be added to the meta.data.

If the Seurat object contains data for multiple runs, a vector containing paths to the VDJ data for each sample can be given. If multiple paths are provided, cell prefixes should be included as names for the vector. 

```{r "import_vdj"}
# Create vector of paths for cellranger output
samples <- levels(so_tcr$orig.ident)
paths   <- file.path("data", str_c(samples, "_TCR"))

names(paths) <- str_c(samples, "_GE")

# Import VDJ data
so_tcr <- import_vdj(
  sobj_in        = so_tcr,  # Seurat object
  vdj_dir        = paths,   # Directories containing cellranger output files
  prefix         = "",      # Prefix to add to new meta.data columns
  filter_contigs = TRUE     # Only include chains with at least one productive contig
)

# Take a look at the meta.data
vdj_cols <- c(
  "clonotype_id", "cdr3",
  "chains", "v_gene", 
  "j_gene", "reads",
  "umis"
)

so_tcr@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

### Quality Control

`djvdj` provides a range of tools to assess data quality and filter single-cell V(D)J data. `mutate_vdj` can be used to add new cell labels to the object meta.data, which is helpful for assessing quality. `filter_vdj` will filter cells based on V(D)J information in the meta.data or just remove V(D)J data from the object without discarding cells. This can be used to filter V(D)J data based on read support. To assess overall read and and UMI support, the function `plot_reads` will summarize these metrics for each clonotype.

```{r "read_support", fig.width = 6.5, fig.height = 3}
plot_reads(
  sobj_in      = so_tcr,        # Seurat object
  chain_col    = "chains",      # Column containing chains for each cell
  cluster_col  = "orig.ident",  # Column containing labels to group by
  plot_colors  = ito_cols       # Plot colors
) +
  guides(fill = FALSE, color = FALSE)
```

<br>

### Clonotype Abundance

To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the `calc_abundance` function. The corresponding `plot_abundance` function will plot clonotype frequency.

```{r "abund_plots", fig.width = 6, fig.height = 3}
plot_abundance(
  sobj_in       = so_tcr,        # Seurat object
  clonotype_col = "cdr3",        # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident",  # meta.data column containing cell labels
  
  plot_colors = ito_cols,        # Plot colors
  yaxis       = "percent",       # Units to plot
  label_col   = "cdr3",          # meta.data column containing labels
  n_labels    = 1,               # Number of top clonotypes to label
  size        = 1                # Additional ggplot options
) +
  theme(legend.title = element_blank())
```

<br>

### Repertoire Diversity

The function `calc_diversity` will calculate repertoire diversity on a per-cluster basis. Using the `cluster_col` argument, any meta.data column containing cell labels can be used for calculations. `calc_diversity` uses the R package `abdiv` for performing diversity calculations and any `abdiv` diversity function can be specified using the `method` argument. The `plot_diversity` function can summarize these results.

```{r "div_plots", fig.width = 7.5, fig.height = 3}
# Metrics to plot
fns <- list(
  "simpson"     = abdiv::simpson,
  "shannon"     = abdiv::shannon,
  "margalef"    = abdiv::margalef,
  "menhinick"   = abdiv::menhinick,
  "brillouin_d" = abdiv::brillouin_d
)

plot_diversity(
  sobj_in       = so_tcr,        # Seurat object
  clonotype_col = "cdr3",        # meta.data column containing clonotype ids
  cluster_col   = "orig.ident",  # meta.data column containing cell labels
  method        = fns,           # abdiv method to use
  plot_colors   = ito_cols
) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

<br>

### Repertoire Overlap

To compare repertoires for different samples or clusters, `calc_similarity` can calculate a variety of different similarity metrics. The `cluster_col` should be used to specify the meta.data column containing cell labels for comparison. Like `calc_diversity`, an `abdiv` function can be specified with the `method` argument. These results can be visualized with `plot_similarity`.

```{r "sim_plots", fig.width = 10, fig.height = 4}
heat_theme <- theme(
  legend.title = element_blank(),
  legend.text  = element_text(size = 8)
)

# Sample heatmap
ident_heat <- plot_similarity(
  sobj_in       = so_tcr,                 # Seurat object
  clonotype_col = "cdr3",                 # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident",           # meta.data column containing cell labels
  method        = abdiv::jaccard,         # Method to use
  plot_colors   = c("grey90", "#009E73")  # Plot colors
) +
  heat_theme

# Cluster heatmap
clust_heat <- plot_similarity(
  sobj_in       = so_tcr,
  clonotype_col = "cdr3",
  cluster_col   = "seurat_clusters",
  method        = abdiv::jaccard,
  plot_colors   = c("grey90", "#56B4E9"),  
  size          = 0.2,                    # Additional ggplot options
  color         = "white"                 # Additional ggplot options
) +
  heat_theme +
  theme(axis.text.x  = element_text(angle = 0))

# Combine heatmaps
plot_grid(ident_heat, clust_heat, align = "h")
```

<br>

### Gene Usage

The V(D)J data imported from Cell Ranger also includes the specific genes detected for each cell. The function `calc_usage` can be used to calculate the fraction of cells that express different V(D)J genes. This function will produce a table summarizing the results. To only include results for a certain chain, the `chain` and `chain_col` arguments can be used to specify the meta.data column containing the chains detected for each cell. These results can be visualized with the `plot_usage` function.

```{r "usage_plots_1", fig.width = 11, fig.height = 3.5}
plot_usage(
  sobj_in     = so_tcr,                # Seurat object
  gene_cols   = "v_gene",              # meta.data column(s) containing genes
  cluster_col = "orig.ident",          # meta.data column containing cell labels
  type        = "bar",                 # Type of plot
  chain       = "TRB",                 # Chain to use for filtering genes
  chain_col   = "chains",              # meta.data column containing chains
  
  yaxis       = "percent",             # Units to plot
  plot_colors = ito_cols,              # Colors to use for heatmap
  plot_genes  = NULL,                  # A list of genes to plot
  n_genes     = NULL,                  # The number of top genes to plot
  
  size        = 0.2,                   # Additional ggplot options
  color       = "white"                # Additional ggplot options
)
```

<br>

By passing multiple columns to `gene_cols`, the frequency that different genes are used together can also be shown.

```{r "usage_plots_2", fig.width = 12, fig.height = 8}
ggs <- plot_usage(
  sobj_in     = so_tcr,                 # Seurat object
  gene_cols   = c("v_gene", "j_gene"),  # meta.data column(s) containing genes
  cluster_col = "orig.ident",           # meta.data column containing cell labels
  chain       = "TRB",                  # Chain to use for filtering genes
  chain_col   = "chains",               # meta.data column containing chains identified
  plot_colors = c("grey90", "#6A51A3")  # Colors to use for heatmap
) %>%
  imap(~ .x + ggtitle(.y))

plot_grid(plotlist = ggs)
```
