---
output: github_document
params:
  data_dir: "~/Projects/Smith_AVIDseq"
  clust_res: 5
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse  = TRUE,
  comment   = "#>",
  fig.path  = "man/figures/README-",
  out.width = "100%",
  warning   = FALSE,
  message   = FALSE
)
```

# djvdj <img src="man/figures/djvdj-logo.png" align="right" height="139">

<!-- badges: start -->
[![Codecov test coverage](https://codecov.io/gh/rnabioco/djvdj/branch/master/graph/badge.svg)](https://codecov.io/gh/rnabioco/djvdj?branch=master)
<!-- badges: end -->

The goal of djvdj is to provide tools to analyze AVID-seq signals alongside single-cell VDJ sequencing data.

<br>

### Installation

You can install the development version of djvdj from [GitHub](https://github.com/rnabioco/djvdj) with:

```{r, eval = F}
# install.packages("devtools")
devtools::install_github("rnabioco/djvdj")
```

```{r "Setup", include = FALSE}
# Packages
library(tidyverse)
library(cowplot)
library(colorblindr)
library(Seurat)
library(here)
library(djvdj)

source(here("funs.R"))

# Themes
base_theme <- theme_cowplot() +
  theme(
    plot.title       = element_text(face = "plain", size = 16),
    strip.background = element_blank(),
    strip.text       = element_text(face = "plain"),
    legend.title     = element_text(size = 10),
    legend.text      = element_text(size = 10),
    axis.title       = element_text(size = 10),
    axis.text        = element_text(size = 10)
  )

umap_theme <- base_theme +
  theme(
    panel.border = element_rect(color = "grey90", size = 1),
    axis.title   = element_blank(),
    axis.line    = element_blank(),
    axis.ticks   = element_blank(),
    axis.text    = element_blank()
  )

# Legend guides
col_guide <- guide_legend(override.aes = list(size = 3.5, shape = 16))

outline_guide <- guide_legend(override.aes = list(
  size   = 3.5,
  shape  = 21,
  color  = "black",
  stroke = 0.25
))

# Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[1:4], "#d7301f", 
  palette_OkabeIto[5:6], "#6A51A3", 
  palette_OkabeIto[7:8], "#875C04"
)

# Set default palette
get_cols <- create_col_fun(ito_cols)
```

<br>

## Vignette

Splenocytes from MD4 transgenic mice which have monoclonal B cells that all bind hen egg lysozyme (HEL) antigen were mixed with splenocytes from C57BL/6 mice at a 1:1 ratio. The cells were stained with the HEL AVID-tag and sequencing libraries were prepared to capture gene expression, B/T cell receptor sequences, and AVID-tag signals using the 10x Genomics 5â€™ immune profiling kit.

```{r "rna_umap", fig.width = 12.5, fig.height = 5, echo = FALSE}
so <- read_rds(here("data/so_1.rds"))

key_gg <- so %>%
  plot_features(
    feature    = "key",
    pt_size    = 0.02,
    pt_outline = 0.2,
    plot_cols  = get_cols()
  ) +
  guides(color = outline_guide) +
  umap_theme +
  theme(
    panel.border = element_blank(),
    legend.title = element_blank()
  )

plot_grid(key_gg, NA)
```

<br>

### Import VDJ data

`import_vdj` takes the output files from `cellranger vdj` and adds clonotype information to the meta.data for an existing Seurat object. For cells that do not have any VDJ sequencing data, NAs will be included in the meta.data.

```{r "import_vdj"}
vdj_dir <- file.path(params$data_dir, "results/JH180_BCR/outs")

so_vdj <- import_vdj(
  sobj_in        = so,       # Seurat object                         
  vdj_dir        = vdj_dir,  # Directory containing cellranger output files
  include_chains = NULL,     # Filter clonotypes that are added based on the given receptor chains
  prefix         = ""        # Prefix to add to new meta.data columns
)

vdj_cols <- c("cdr3s_aa", "n_chains", "clone_freq", "clone_frac")

so_vdj@meta.data %>%
  as_tibble() %>%
  select(orig.ident, nCount_RNA, nFeature_RNA, all_of(vdj_cols))
```

<br>

### Filtering

`filter_vdj` allows you to filter a Seurat object using the added clonotype information or any other columns present in the meta.data. The terms `.chains` and `.seqs` can be used to filter based on the chains detected for the cell or the sequence of the chain. Filtering is only performed on cells that include VDJ data, all other cells will remain in the object.

Filter to only include cells with both IGH and IGK chains

```{r}
so_filt <- filter_vdj(
  sobj_in  = so_vdj,                  # Seurat object
  cdr3_col = "cdr3s_aa",              # meta.data column containing CDR3 sequences
  all(c("IGH", "IGK") %in% .chains),  # Expression to use for filtering
)

so_filt@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

Filter for cells with an IGL chain and at least two cells or >1% that share the clonotype

```{r}
so_filt <- so_vdj %>%
  filter_vdj(.chains == "IGL" && clone_freq > 1 || clone_frac > 0.01)

so_filt@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

Other examples

```{r, eval = FALSE}
# Cells with CDR3 amino acid sequence of 'CQQSNSWPYTF'
so_filt <- so_vdj %>%
  filter_vdj(.seqs == 'CQQSNSWPYTF')

# Cells with IGH, IGK, and IGL chains
so_filt <- so_vdj %>%
  filter_vdj(all(c("IGH", "IGK", "IGL") %in% .chains))

# Cells with two IGK chains and no other chains and >1000 RNA counts
so_filt <- so_vdj %>%
  filter_vdj(all(.chains == "IGK") && n_chains == 2 && nCount_RNA > 1000)

# Cells with at least two unique chains and no IGH chain
so_filt <- so_vdj %>%
  filter_vdj(all(.chains != "IGH") && length(unique(.chains)) > 1)
```

<br>

### Clustering

Cells can be clustered based on the Levenshtein distance between CDR3 sequences using `cluster_vdj`. The `use_chains` argument can be used to select the chains that should be used when performing distance calculations.

```{r}
so_vdj <- cluster_vdj(
  sobj_in    = so_vdj,            # Seurat object
  cdr3_col   = "cdr3s_aa",        # meta.data column containing CDR3 sequences
  resolution = params$clust_res,  # Clustering resolution
  use_chains = NULL,              # Chains to use for distance calculations
  prefix     = "vdj_"             # Prefix to add to new graph
)

so_vdj@meta.data %>%
  as_tibble() %>%
  select(all_of(vdj_cols), seurat_clusters)
```

<br>

The Seurat `RunUMAP` function will not run with the hybrid VDJ object, but the wrapper function `run_umap_vdj` can be used.

```{r}
so_vdj <- run_umap_vdj(
  sobj_in   = so_vdj,      # Seurat object
  umap_key  = "vdjUMAP_",  # Prefix for UMAP columns
  vdj_graph = "vdj_snn"    # Name of VDJ graph to use for UMAP
)
```

```{r "vdj_umap", fig.width = 11, fig.height = 5, echo = FALSE}
rna_gg <- so_vdj %>%
  plot_features(
    feature    = "seurat_clusters",
    pt_size    = 0.02,
    pt_outline = 0.2,
    plot_cols  = get_cols()
  ) +
  ggtitle("RNA") +
  guides(color = outline_guide) +
  umap_theme +
  theme(legend.title = element_blank())

gg_legd <- get_legend(rna_gg)

rna_gg <- rna_gg +
  theme(legend.position = "none")

vdj_gg <- so_vdj %>%
  plot_features(
    x          = "vdjUMAP_1",
    y          = "vdjUMAP_2",
    feature    = "seurat_clusters",
    pt_size    = 0.02,
    pt_outline = 0.2,
    plot_cols  = get_cols()
  ) +
  ggtitle("VDJ") +
  umap_theme +
  theme(legend.position = "none")

plot_grid(
  rna_gg, vdj_gg, gg_legd,
  rel_widths = c(1, 1, 0.2),
  nrow       = 1
)
```

<br>

### Repertoire stats

The functions `calc_diversity` and `calc_jaccard` will calculate repertoire diversity and repertoire overlap on a per-cluster basis. These functions can be given any meta.data column containing cell labels to use for calculations.

Calculate repertoire diversity with `calc_diversity`. The inverse Simpson index is used to measure diversity for each cluster.

```{r}
so_vdj <- calc_diversity(
  sobj_in       = so_vdj,          # Seurat object
  clonotype_col = "clonotype_id",  # meta.data column containing clonotype ids
  cluster_col   = NULL,            # meta.data column containing cell labels
  prefix        = ""               # Prefix to add to new meta.data columns
)
```

```{r "calc_diversity", fig.width = 12, fig.height = 5, echo = FALSE}
div_gg <- so_vdj %>%
  plot_features(
    x          = "vdjUMAP_1",
    y          = "vdjUMAP_2",
    feature    = "diversity",
    pt_size    = 0.02,
    pt_outline = 0.2,
    plot_cols  = c("grey90", get_cols()[1])
  ) +
  ggtitle("Repertoire Diversity") +
  guides(color = guide_colorbar(barwidth = unit(5, "pt"))) +
  umap_theme +
  theme(legend.title = element_blank())

hel_gg <- so_vdj %>%
  plot_features(
    x          = "vdjUMAP_1",
    y          = "vdjUMAP_2",
    feature    = "adt_-HEL",
    pt_size    = 0.02,
    pt_outline = 0.2,
    plot_cols  = c("grey90", get_cols()[6])
  ) +
  ggtitle("HEL") +
  guides(color = guide_colorbar(barwidth = unit(5, "pt"))) +
  umap_theme +
  theme(legend.title = element_blank())

plot_grid(
  div_gg, hel_gg,
  align  = "vh",
  "axis" = "trbl"
)
```

<br>

Calculate repertoire overlap with `calc_jaccard`. A reference cell label can be given to calculate Jaccard index for the reference vs all other cell groups. If no reference is given overlap is calculated for all combinations of cell labels.

```{r}
so_vdj <- calc_jaccard(
  sobj_in       = so_vdj,             # Seurat object
  clonotype_col = "clonotype_id",     # meta.data column containing clonotype ids
  cluster_col   = "seurat_clusters",  # meta.data column containing cell labels
  ref_cluster   = NULL,               # Cell label to use as a reference for Jaccard index
  prefix        = "x"                 # Prefix to add to new meta.data columns 
)
```

```{r "calc_jaccard", fig.width = 12.5, fig.height = 5, echo = FALSE}
j_gg <- so_vdj %>%
  plot_features(
    x          = "vdjUMAP_1",
    y          = "vdjUMAP_2",
    feature    = "x8_jaccard",
    pt_size    = 0.02,
    pt_outline = 0.2,
    plot_cols  = c("grey90", get_cols()[2])
  ) +
  ggtitle("Repertoire Overlap") +
  guides(color = guide_colorbar(barwidth = unit(5, "pt"))) +
  umap_theme

plot_grid(j_gg, NA)
```

<br>

### Combine with TCR data

BCR and TCR data can be added to the same Seurat object

```{r}
vdj_dir <- file.path(params$data_dir, "results/JH180_TCR/outs")

so_vdj <- so_vdj %>%
  import_vdj(
    vdj_dir = vdj_dir,
    prefix  = "tcr_"
  ) %>%
  calc_diversity(
    clonotype_col = "tcr_clonotype_id",
    cluster_col   = "RNA_clusters",
    prefix        = "tcr_"
  )
```

```{r "bcr_tcr", fig.width = 10, fig.height = 5.5, echo = F}
key_gg <- so %>%
  plot_features(
    feature    = "key",
    pt_size    = 0.2,
    pt_outline = 0.3,
    plot_cols  = get_cols()
  ) +
  guides(color = outline_guide) +
  umap_theme +
  theme(
    legend.position = "bottom",
    legend.title = element_blank()
  )

div_gg <- so_vdj@meta.data %>%
  mutate(
    merged_div = diversity,
    merged_div = ifelse(is.na(diversity), tcr_diversity * -1, diversity),
  ) %>%
  filter(!is.na(merged_div)) %>%
  plot_features(
    feature    = "merged_div",
    plot_cols  = c(get_cols()[7], "white", get_cols()[8]),
    pt_size    = 0.2, 
    pt_outline = 0.3
  ) +
  ggtitle("BCR/TCR Diversity") +
  guides(color = guide_colorbar(barheight = unit(5, "pt"))) +
  umap_theme +
  theme(
    legend.position = "bottom",
    legend.title = element_blank()
  )

plot_grid(
  key_gg, div_gg,
  align = "vh",
  axis  = "trbl"
)
```


