<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Collection of Tools to Analyze AVID-seq Data • djvdj</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Collection of Tools to Analyze AVID-seq Data">
<meta property="og:description" content="What the package does (one paragraph).">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">djvdj</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/bcr.html">BCR analysis</a>
    </li>
    <li>
      <a href="articles/tcr.html">TCR analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="djvdj-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#djvdj-" class="anchor"></a>djvdj <img src="reference/figures/djvdj-logo.png" align="right" height="145">
</h1></div>
<!-- badges: start -->

<p>The goal of djvdj is to provide tools to analyze AVID-seq signals alongside single-cell VDJ sequencing data.</p>
<p><br></p>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<p>You can install the development version of djvdj from <a href="https://github.com/rnabioco/djvdj">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"rnabioco/djvdj"</span><span class="op">)</span></pre></div>
<p><br></p>
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. In nec molestie risus. Duis vitae eros odio. Proin condimentum odio dolor, at vehicula lorem tempus sit amet. Nunc id metus vehicula, facilisis velit in, tincidunt augue. Donec at semper turpis, nec cursus justo. Suspendisse est lorem, vulputate vitae diam sit amet, luctus volutpat sem. Aenean at odio sed nibh tempor eleifend. Duis ultrices turpis sit amet velit luctus, eu commodo urna lacinia. In porttitor tristique quam ac molestie. In s</p>
<p><img src="reference/figures/README-rna_umap-1.png"><!-- --></p>
<p><br></p>
</div>
<div id="import-vdj-data" class="section level3">
<h3 class="hasAnchor">
<a href="#import-vdj-data" class="anchor"></a>Import VDJ data</h3>
<p><code>import_vdj</code> takes the output files from <code>cellranger vdj</code> and adds clonotype information to the meta.data for an existing Seurat object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a “;” (or a character specified by <code>sep</code>). For cells that do not have any VDJ sequencing data, NAs will be added to the meta.data.</p>
<p>If the Seurat object contains data for multiple runs, a vector containing paths to the VDJ data for each sample can be given. If multiple paths are provided, cell prefixes should be included as names for the vector.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># Create vector of paths for cellranger output</span>
<span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">so_tcr</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span>
<span class="va">paths</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="fu">str_c</span><span class="op">(</span><span class="va">samples</span>, <span class="st">"_TCR"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">str_c</span><span class="op">(</span><span class="va">samples</span>, <span class="st">"_GE"</span><span class="op">)</span>

<span class="co"># Import VDJ data</span>
<span class="va">so_tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>
  sobj_in        <span class="op">=</span> <span class="va">so_tcr</span>,  <span class="co"># Seurat object</span>
  vdj_dir        <span class="op">=</span> <span class="va">paths</span>,   <span class="co"># Directories containing cellranger output files</span>
  prefix         <span class="op">=</span> <span class="st">""</span>,      <span class="co"># Prefix to add to new meta.data columns</span>
  filter_contigs <span class="op">=</span> <span class="cn">TRUE</span>     <span class="co"># Only include chains with at least one productive contig</span>
<span class="op">)</span></pre></div>
<p><br></p>
</div>
<div id="quality-control" class="section level3">
<h3 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality Control</h3>
<p><code>djvdj</code> provides a range of tools to assess data quality and filter single-cell V(D)J data. <code>mutate_vdj</code> can be used to add new cell labels to the object meta.data, which is helpful for assessing quality. <code>filter_vdj</code> will filter cells based on V(D)J information in the meta.data or just remove V(D)J data from the object without discarding cells. This can be used to filter V(D)J data based on read support. To assess overall read and and UMI support, the function <code>plot_reads</code> will summarize these metrics for each clonotype.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="reference/plot_reads.html">plot_reads</a></span><span class="op">(</span>
  sobj_in      <span class="op">=</span> <span class="va">so_tcr</span>,        <span class="co"># Seurat object</span>
  chain_col    <span class="op">=</span> <span class="st">"chains"</span>,      <span class="co"># Column containing chains for each cell</span>
  cluster_col  <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># Column containing labels to group by</span>
  plot_colors  <span class="op">=</span> <span class="va">ito_cols</span>       <span class="co"># Plot colors</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/README-read_support-1.png"><!-- --></p>
<p><br></p>
</div>
<div id="clonotype-abundance" class="section level3">
<h3 class="hasAnchor">
<a href="#clonotype-abundance" class="anchor"></a>Clonotype Abundance</h3>
<p>To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the <code>calc_abundance</code> function. The corresponding <code>plot_abundance</code> function will plot clonotype frequency.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="reference/plot_abundance.html">plot_abundance</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_tcr</span>,        <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,        <span class="co"># meta.data column containing clonotype IDs</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># meta.data column containing cell labels</span>
  
  plot_colors <span class="op">=</span> <span class="va">ito_cols</span>,        <span class="co"># Plot colors</span>
  yaxis       <span class="op">=</span> <span class="st">"percent"</span>,       <span class="co"># Units to plot</span>
  label_col   <span class="op">=</span> <span class="st">"cdr3"</span>,          <span class="co"># meta.data column containing labels</span>
  n_labels    <span class="op">=</span> <span class="fl">1</span>,               <span class="co"># Number of top clonotypes to label</span>
  size        <span class="op">=</span> <span class="fl">1</span>                <span class="co"># Additional ggplot options</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/README-abund_plots-1.png"><!-- --></p>
<p><br></p>
</div>
<div id="repertoire-diversity" class="section level3">
<h3 class="hasAnchor">
<a href="#repertoire-diversity" class="anchor"></a>Repertoire Diversity</h3>
<p>The function <code>calc_diversity</code> will calculate repertoire diversity on a per-cluster basis. Using the <code>cluster_col</code> argument, any meta.data column containing cell labels can be used for calculations. <code>calc_diversity</code> uses the R package <code>abdiv</code> for performing diversity calculations and any <code>abdiv</code> diversity function can be specified using the <code>method</code> argument. The <code>plot_diversity</code> function can summarize these results.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># Metrics to plot</span>
<span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
  <span class="st">"simpson"</span>     <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/simpson.html">simpson</a></span>,
  <span class="st">"shannon"</span>     <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">shannon</a></span>,
  <span class="st">"margalef"</span>    <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/margalef.html">margalef</a></span>,
  <span class="st">"menhinick"</span>   <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/menhinick.html">menhinick</a></span>,
  <span class="st">"brillouin_d"</span> <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">brillouin_d</a></span>
<span class="op">)</span>

<span class="fu"><a href="reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_tcr</span>,        <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,        <span class="co"># meta.data column containing clonotype ids</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># meta.data column containing cell labels</span>
  method        <span class="op">=</span> <span class="va">fns</span>,           <span class="co"># abdiv method to use</span>
  plot_colors   <span class="op">=</span> <span class="va">ito_cols</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/README-div_plots-1.png"><!-- --></p>
<p><br></p>
</div>
<div id="repertoire-overlap" class="section level3">
<h3 class="hasAnchor">
<a href="#repertoire-overlap" class="anchor"></a>Repertoire Overlap</h3>
<p>To compare repertoires for different samples or clusters, <code>calc_similarity</code> can calculate a variety of different similarity metrics. The <code>cluster_col</code> should be used to specify the meta.data column containing cell labels for comparison. Like <code>calc_diversity</code>, an <code>abdiv</code> function can be specified with the <code>method</code> argument. These results can be visualized with <code>plot_similarity</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">heat_theme</span> <span class="op">&lt;-</span> <span class="fu">theme</span><span class="op">(</span>
  legend.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
  legend.text  <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># Sample heatmap</span>
<span class="va">ident_heat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_tcr</span>,                 <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,                 <span class="co"># meta.data column containing clonotype IDs</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,           <span class="co"># meta.data column containing cell labels</span>
  method        <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>,         <span class="co"># Method to use</span>
  plot_colors   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey90"</span>, <span class="st">"#009E73"</span><span class="op">)</span>  <span class="co"># Plot colors</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="va">heat_theme</span>

<span class="co"># Cluster heatmap</span>
<span class="va">clust_heat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_tcr</span>,
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,
  cluster_col   <span class="op">=</span> <span class="st">"seurat_clusters"</span>,
  method        <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>,
  plot_colors   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey90"</span>, <span class="st">"#56B4E9"</span><span class="op">)</span>,  
  size          <span class="op">=</span> <span class="fl">0.2</span>,                    <span class="co"># Additional ggplot options</span>
  color         <span class="op">=</span> <span class="st">"white"</span>                 <span class="co"># Additional ggplot options</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="va">heat_theme</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text.x  <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Combine heatmaps</span>
<span class="fu">plot_grid</span><span class="op">(</span><span class="va">ident_heat</span>, <span class="va">clust_heat</span>, align <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/README-sim_plots-1.png"><!-- --></p>
<p><br></p>
</div>
<div id="gene-usage" class="section level3">
<h3 class="hasAnchor">
<a href="#gene-usage" class="anchor"></a>Gene Usage</h3>
<p>The V(D)J data imported from Cell Ranger also includes the specific genes detected for each cell. The function <code>calc_usage</code> can be used to calculate the fraction of cells that express different V(D)J genes. This function will produce a table summarizing the results. To only include results for a certain chain, the <code>chain</code> and <code>chain_col</code> arguments can be used to specify the meta.data column containing the chains detected for each cell. These results can be visualized with the <code>plot_usage</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="reference/plot_usage.html">plot_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,              <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,          <span class="co"># meta.data column containing cell labels</span>
  type        <span class="op">=</span> <span class="st">"bar"</span>,                 <span class="co"># Type of plot</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,                 <span class="co"># Chain to use for filtering genes</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>,              <span class="co"># meta.data column containing chains</span>
  
  yaxis       <span class="op">=</span> <span class="st">"percent"</span>,             <span class="co"># Units to plot</span>
  plot_colors <span class="op">=</span> <span class="va">ito_cols</span>,              <span class="co"># Colors to use for heatmap</span>
  plot_genes  <span class="op">=</span> <span class="cn">NULL</span>,                  <span class="co"># A list of genes to plot</span>
  n_genes     <span class="op">=</span> <span class="cn">NULL</span>,                  <span class="co"># The number of top genes to plot</span>
  
  size        <span class="op">=</span> <span class="fl">0.2</span>,                   <span class="co"># Additional ggplot options</span>
  color       <span class="op">=</span> <span class="st">"white"</span>                <span class="co"># Additional ggplot options</span>
<span class="op">)</span></pre></div>
<p><img src="reference/figures/README-usage_plots_1-1.png"><!-- --></p>
<p><br></p>
<p>By passing multiple columns to <code>gene_cols</code>, the frequency that different genes are used together can also be shown.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">ggs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_usage.html">plot_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                 <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,  <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,           <span class="co"># meta.data column containing cell labels</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,                  <span class="co"># Chain to use for filtering genes</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>,               <span class="co"># meta.data column containing chains identified</span>
  plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey90"</span>, <span class="st">"#6A51A3"</span><span class="op">)</span>  <span class="co"># Colors to use for heatmap</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">imap</span><span class="op">(</span><span class="op">~</span> <span class="va">.x</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="va">.y</span><span class="op">)</span><span class="op">)</span>

<span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">ggs</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/README-usage_plots_2-1.png"><!-- --></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ryan Sheridan <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/YOUR-ORCID-ID" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/rnabioco/djvdj/actions"><img src="https://github.com/rnabioco/djvdj/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Ryan Sheridan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
