---
title: "TCR analysis"
date: '`r format(Sys.Date(), "%B %d %Y")`'
output:
  rmarkdown::html_vignette:
    toc_float: true
    toc_depth: 2
    vignette: >
      %\VignetteIndexEntry{TCR analysis}
      %\VignetteEncoding{UTF-8}
      %\VignetteEngine{knitr::rmarkdown}
---

```{r "setup", include = FALSE}
# Chunk opts
knitr::opts_chunk$set(
  collapse  = TRUE,
  comment   = "#>",
  warning   = FALSE,
  message   = FALSE
)

# Packages
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(ggplot2)
library(cowplot)
library(colorblindr)
library(here)
library(Seurat)
library(djvdj)

# Okabe Ito color palettes
vdj_colors <- c(
  palette_OkabeIto[1:4], "#d7301f", 
  palette_OkabeIto[5:6], "#6A51A3", 
  palette_OkabeIto[7:8], "#875C04"
)
```

<br>

## V(D)J Sequencing

For 10x single-cell immune profiling, the initial steps of GEM formation and reverse transcription are the same. After reverse transcription within each GEM, a portion of the pooled cDNA is further amplified using a primer specific for the V(D)J constant region. The amplified cDNA is then fragmented which results in fragments of various length that span the entire gene. The remaining cDNA that was not amplified using V(D)J specific primers is used to generate the gene expression library. This ultimately results in separate gene expression and V(D)J libraries that include the same cell barcodes.

<br>

<img src="10x_GEM_generation.png" width="400"> <img src="10x_GEM_vdj_cDNA.png" width="400"> 

<br>

Single-cell V(D)J libraries are typically sequenced at a depth of ~10k reads per cell. After sequencing, the raw fastq files are processed separately from the gene expression libraries using the [Cell Ranger](https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/vdj) software available through 10x Genomics.

<br>

<img src="10x_vdj_workflow.png" width="800"> 

<br>

When analyzing the V(D)J results alongside single-cell gene expression data, there are several key differences to keep in mind. The main goal of V(D)J sequencing is to accurately assemble each chain expressed in the cell, not to quantify changes in expression. Due to this difference, the total number of UMIs obtained for each cell can be much lower than the gene expression libraries. In addition, when processing gene expression data, reads with the same UMI are normally discarded. However, for V(D)J sequencing these reads are kept and are important for obtaining sufficient coverage to accurately assemble each chain.

<br>

<img src="10x_vdj_read_layout.png" width="800"> 

<br>

Lorem ipsum dolor sit amet, consectetur adipiscing elit. In nec molestie risus. Duis vitae eros odio. Proin condimentum odio dolor, at vehicula lorem tempus sit amet. Nunc id metus vehicula, facilisis velit in, tincidunt augue. Donec at semper turpis, nec cursus justo. Suspendisse est lorem, vulputate vitae diam sit amet, luctus volutpat sem. Aenean at odio sed nibh tempor eleifend. Duis ultrices turpis sit amet velit luctus, eu commodo urna lacinia. In porttitor tristique quam ac molestie. In s

```{r "rna_umap", fig.width = 12.5, fig.height = 5, echo = FALSE}
load(here("data/tcr/so_tcr.rda"))

# Sample UMAP
key_umap <- so_tcr %>%
  DimPlot(
    group.by = "orig.ident",
    cols     = vdj_colors
  ) +
  ggtitle("Sample") +
  theme_nothing() +
  theme(
    plot.title = element_text(),
    legend.position = "right"
  )

# Cluster UMAP
clust_umap <- so_tcr %>%
  DimPlot(group.by = "seurat_clusters") +
  ggtitle("Cluster") +
  theme_nothing() +
  theme(
    plot.title = element_text(),
    legend.position = "right"
  )

# Final figure
plot_grid(
  key_umap, clust_umap,
  rel_widths = c(1, 0.9)
)
```

<br>

## Import VDJ data

`import_vdj()` takes the output files from [Cell Ranger](https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/vdj#header) and adds clonotype information to the meta.data for an existing [Seurat](https://satijalab.org/seurat/) object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a ';' (or a character specified by `sep`). For cells that do not have any VDJ sequencing data, `NA`s will be added to the meta.data.

If the Seurat object contains data for multiple runs, a vector containing paths to the VDJ data for each sample can be given. If multiple paths are provided, cell prefixes should be included as names for the vector. 

```{r "import_vdj"}
# Create vector of paths pointing to cellranger output
paths <- c(
  KI_DN3_GE = "../data/tcr/KI_DN3_TCR",
  KI_DN4_GE = "../data/tcr/KI_DN4_TCR",
  WT_DN3_GE = "../data/tcr/WT_DN3_TCR",
  WT_DN4_GE = "../data/tcr/WT_DN4_TCR"
)

so_tcr <- import_vdj(
  sobj_in        = so_tcr,                  # Seurat object
  vdj_dir        = paths,                   # cellranger directories
  prefix         = "",                      # Prefix to add to new meta.data columns
  filter_contigs = TRUE                     # Only include productive chains
)
```
<br>

Take a look at the meta.data to see the V(D)J data added to the object.

```{r}
vdj_cols <- c(
  "clonotype_id", "cdr3",
  "chains", "v_gene", 
  "j_gene", "reads",
  "umis"
)

so_tcr@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

<br>

## Quality Control

### Read Support

The read support for each chain can be visualized with the `plot_reads()` function. This will create plots summarizing the number of UMIs and total reads that were obtained for each chain.

```{r "read_support", fig.width = 6.5, fig.height = 3}
plot_reads(
  sobj_in     = so_tcr,                     # Seurat object
  chain_col   = "chains",                   # Column containing chains for each cell
  cluster_col = "orig.ident",               # Column containing labels to group by
  plot_colors = vdj_colors
) +
  guides(fill = FALSE, color = FALSE)
```

<br>

The djvdj package provides several functions that can be used to manipulate the object meta.data, which is useful for identifying cells that should be removed. This includes a function to modify meta.data columns (`mutate_vdj()`) and a function to filter cells from the object (`filter_vdj()`). Since cells can have multiple values present in each V(D)J column (separated by a semicolon), when using these functions each string of V(D)J values for a cell should be thought of as a vector.

For example using `filter_vdj()` we can remove V(D)J data for cells that have minimal UMI support. To do this we use the expression `all(umis > 2)`{.R}. This means that all semicolon-separated values present in the `umis` column must be greater than 2. Setting `filter_cells = FALSE`{.R} means the cell will remain in the object, but the V(D)J data will be removed.

```{r}
so_tcr <- filter_vdj(
  sobj_in      = so_tcr,                    # Seurat object
  filt         = all(umis > 2),             # Condition for filtering
  filter_cells = FALSE                      # Should cells be removed from object
)

# Take a look at the meta.data
so_tcr@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id)) %>%
  select(all_of(vdj_cols))
```

To further illustrate how `filter_vdj()` works we can look at row 7 shown above, which has two beta chains with 7 UMIs for the first and 171 UMIs for the second. When filtering this row we are evaluating whether the following condition is `TRUE`:

```{r}
vec <- c(7, 171)

all(vec > 2)
```

<br>

### Paired Chains

The V(D)J data imported from Cell Ranger will include clonotypes that do not have paired alpha and beta chains. Using the `mutate_vdj()` function, we can add a new cell label to the meta.data to allow for easy visualization of these cells.

In this example we create a new column (`uniq_chains`) that summarizes the unique chains identified for each cell. We can then overlay these cell labels on a UMAP projection.

```{r}
so_tcr <- mutate_vdj(
  sobj_in     = so_tcr,                                # Seurat object
  uniq_chains = str_c(unique(chains), collapse = "_")  # New column
)

# Take a look at the meta.data
so_tcr@meta.data %>%
  as_tibble() %>%
  filter(!is.na(clonotype_id), n_chains > 2) %>%
  select(all_of(vdj_cols), uniq_chains)
```

For row 1 shown above, this is equivalent to:

```{r}
vec <- c("TRA", "TRB", "TRB")
vec <- unique(vec)

str_c(vec, collapse = "_")
```

<br>

We can then overlay the new labels present in the `uniq_chains` column on a UMAP projection. djvdj also provides a simple wrapper function (`plot_cell_count()`) that will summarize the fraction of cells that belong to each group.

```{r "chains_umap", fig.width = 6.5, fig.height = 4.2}
chain_cols <- vdj_colors
names(chain_cols) <- unique(so_tcr$uniq_chains)

# Chains UMAP
chain_umap <- so_tcr %>%
  DimPlot(
    group.by = "uniq_chains",
    cols     = chain_cols,
    na.value = "grey90"
  ) +
  theme_nothing() +
  theme(legend.position = "top")
  
# Chains bar graphs
chain_bars <- so_tcr %>%
  plot_cell_count(
    x           = "orig.ident",             # Cell labels to plot
    fill_col    = "uniq_chains",            # Cell labels to color each bar
    plot_colors = chain_cols
  ) +
  theme(
    plot.margin     = unit(c(120, 10, 10, 10), "pt"),
    legend.position = "none"
  )

# Final figure
plot_grid(
  chain_umap, chain_bars,
  rel_widths = c(1, 0.55),
  nrow       = 1
)
```

<br>

Clonotypes that lack paired alpha and beta chains can be filtered using `filter_vdj()`. However, for this dataset we expect a large fraction of cells to have unpaired chains, so we can skip this step.

```{r}
so_filt <- filter_vdj(
  sobj_in      = so_tcr,                            # Seurat object
  filt         = all(c("TRA", "TRB") %in% chains),  # Condition for filtering
  filter_cells = FALSE                              # Should cells be removed
)
```

<br>

## Clonotype Abundance

To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the `calc_abundance()` function. These calculations can be performed on a per-cluster or per-sample basis by also providing a meta.data column containing cell labels.

```{r}
so_tcr <- calc_abundance(
  sobj_in       = so_tcr,                   # Seurat object
  clonotype_col = "cdr3_nt",                # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident"              # meta.data column containing cell labels
)
```

<br>

For each 'calc' function provided by djvdj, there is a matching 'plot' function that will generate a summary plot. The `plot_abundance()` function will plot clonotypes ranked by abundance.

```{r "abund_plots", fig.width = 10, fig.height = 4}
plot_abundance(
  sobj_in       = so_tcr,                   # Seurat object
  clonotype_col = "cdr3_nt",                # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident",             # meta.data column containing cell labels
  label_col     = "cdr3",                   # meta.data column containing labels
  n_clonotypes  = 12,                       # Number of top clonotypes
  plot_colors   = vdj_colors
) +
  theme(
    legend.position = "none",
    axis.text.x     = element_text(size = 6)
  )
```

<br>

## Repertoire Diversity

The function `calc_diversity()` will calculate repertoire diversity on number of cells that share each clonotype. Using the `cluster_col` argument, any meta.data column containing cell labels can be used for calculations. `calc_diversity()` uses the R package [abdiv](https://github.com/kylebittinger/abdiv) for performing diversity calculations and any [abdiv](https://github.com/kylebittinger/abdiv) diversity function can be specified using the `method` argument.

Possible methods for calculating diversity include:

```{r, comment = "", echo = F}
abdiv::alpha_diversities
```
<br>

In this example we are calculating the Shannon diversity for each sample in the orig.ident meta.data column.

```{r}
so_tcr <- calc_diversity(
  sobj_in       = so_tcr,                   # Seurat object
  clonotype_col = "cdr3_nt",                # meta.data column containing clonotype ids
  cluster_col   = "orig.ident",             # meta.data column containing cell labels
  method        = abdiv::shannon            # abdiv method to use
)
```

<br>

The `plot_diversity()` function will create plots summarizing repertoire diversity for each sample. A named list of functions to run can also be passed to plot multiple metrics. Four different diversity metrics are shown in the example below.

```{r "div_plots", fig.width = 7.5, fig.height = 3}
# Metrics to plot
fns <- list(
  "simpson"     = abdiv::simpson,
  "shannon"     = abdiv::shannon,
  "margalef"    = abdiv::margalef,
  "menhinick"   = abdiv::menhinick,
  "brillouin_d" = abdiv::brillouin_d
)

plot_diversity(
  sobj_in       = so_tcr,                   # Seurat object
  clonotype_col = "cdr3_nt",                # meta.data column containing clonotype ids
  cluster_col   = "orig.ident",             # meta.data column containing cell labels
  method        = fns,                      # abdiv method to use
  plot_colors   = vdj_colors
) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

<br>

## Repertoire Overlap

To compare repertoires for different samples or clusters, `calc_similarity()` can calculate a variety of different similarity metrics. The `cluster_col` should be used to specify the meta.data column containing cell labels for comparison. Like `calc_diversity()`, an [abdiv](https://github.com/kylebittinger/abdiv) function can be specified with the `method` argument.

Possible methods for calculating repertoire similarity include:

```{r, comment = "", echo = F}
abdiv::beta_diversities
```

<br>

By default `calc_similarity()` will add a new meta.data column for each comparison. In this example we are calculating the jaccard dissimilarity index for all combinations of cell labels present in the `orig.ident` column.

```{r}
so_tcr <- calc_similarity(
  sobj_in       = so_tcr,                   # Seurat object
  clonotype_col = "cdr3_nt",                # meta.data column containing clonotype ids
  cluster_col   = "orig.ident",             # meta.data column containing cell labels
  method        = abdiv::jaccard            # abdiv method to use
)
```

<br>

A heatmap summarizing the results can be generated using the `plot_similarity()` function. Here we are creating two heatmaps, one to compare the different samples and one to compare cell clusters.

```{r "sim_plots", fig.width = 10, fig.height = 4}

# Sample heatmap
ident_heat <- plot_similarity(
  sobj_in       = so_tcr,                   # Seurat object
  clonotype_col = "cdr3_nt",                # meta.data column containing clonotype IDs
  cluster_col   = "orig.ident",             # meta.data column containing cell labels
  method        = abdiv::jaccard,           # Method to use
  plot_colors   = "#009E73"
) +
  theme(legend.text = element_text(size = 8))

# Cluster heatmap
clust_heat <- plot_similarity(
  sobj_in       = so_tcr,
  clonotype_col = "cdr3_nt",
  cluster_col   = "seurat_clusters",
  method        = abdiv::jaccard,
  plot_colors   = "#56B4E9",
  size          = 0.2,                      # Additional ggplot options
  color         = "white"                   # Additional ggplot options
) +
  theme(
    legend.text = element_text(size = 8),
    axis.text.x = element_text(angle = 0)
  )

# Combine heatmaps
plot_grid(ident_heat, clust_heat, align = "h")
```

<br>

## Gene Usage

The V(D)J data imported from Cell Ranger also includes the specific genes detected for each cell. The function `calc_usage()` can be used to calculate the fraction of cells that express different V(D)J genes. This function will produce a table summarizing the results. To only include results for a certain chain, the `chain` and `chain_col` arguments can be used to specify the meta.data column containing the chains detected for each cell. By default the results for all chains will be included.

In this example we are summarizing the usage of different V genes for the TRB chain 

```{r}
calc_usage(
  sobj_in     = so_tcr,                     # Seurat object
  gene_cols   = "v_gene",                   # meta.data column containing genes
  cluster_col = "orig.ident",               # meta.data column containing cell labels
  chain       = "TRB",                      # Chain to use for filtering genes
  chain_col   = "chains"                    # meta.data column containing chains
)
```

<br>

The function `plot_usage()` can be used to summarize these results. Using the `yaxis` argument, the percent or absolute count (frequency) can be used for plotting. The genes plotted can also be selected using the `plot_genes` argument, or the number of top genes (most frequent) to plot can be specified with `n_genes`.

```{r "gene_usage_1", fig.width = 11, fig.height = 3.5}
plot_usage(
  sobj_in     = so_tcr,                     # Seurat object
  gene_cols   = "v_gene",                   # meta.data column(s) containing genes
  cluster_col = "orig.ident",               # meta.data column containing cell labels
  type        = "bar",                      # Type of plot
  chain       = "TRB",                      # Chain to use for filtering genes
  chain_col   = "chains",                   # meta.data column containing chains
  
  plot_genes  = NULL,                       # A list of genes to plot
  n_genes     = NULL,                       # The number of top genes to plot
  plot_colors = vdj_colors
)
```

<br>

By passing multiple columns to `gene_cols`, the frequency that different genes are used together can also be summarized.

```{r, eval = F}
calc_usage(
  sobj_in     = so_tcr,                     # Seurat object
  gene_cols   = c("v_gene", "j_gene"),      # meta.data column(s) containing genes
  cluster_col = "orig.ident",               # meta.data column containing cell labels
  chain       = "TRB",                      # Chain to use for filtering genes
  chain_col   = "chains"                    # meta.data column containing chains
)
```

<br>

When multiple gene columns are passed to `plot_usage()`, a list of plots will be returned, one for each cell label in the `cluster_col` column.

```{r "gene_usage_2", fig.width = 12, fig.height = 8}
ggs <- plot_usage(
  sobj_in     = so_tcr,                     # Seurat object
  gene_cols   = c("v_gene", "j_gene"),      # meta.data column(s) containing genes
  cluster_col = "orig.ident",               # meta.data column containing cell labels
  chain       = "TRB",                      # Chain to use for filtering genes
  chain_col   = "chains",                   # meta.data column containing chains identified
  plot_colors = "#6A51A3"
) %>%
  imap(~ .x + ggtitle(.y))

plot_grid(plotlist = ggs)
```

<br>

## CDR3 Usage

The `calc_usage()` and `plot_usage()` functions can also be used to calculate the fraction of cells that have a certain CDR3 sequence. To do this, the meta.data column containing the CDR3 sequences for each cell can be passed to the `gene_cols` argument. The plot can be divided into facets based on sample using simple ggplot2 commands.

```{r "cdr3_usage", fig.width = 10, fig.height = 4}
plot_usage(
  sobj_in     = so_tcr,
  gene_cols   = "cdr3",                     # Column containing CDR3 sequences
  cluster_col = "orig.ident",
  
  type        = "bar",
  chain       = "TRB",
  chain_col   = "chains",
  plot_colors = vdj_colors,
  n_genes     = 12                          # Number of CDR3 sequences to plot
) +
  facet_wrap(~ orig.ident, nrow = 1, scales = "free_x") +
  theme(legend.position = "none")
```

```{r "DIM RED EX", eval = F, echo = F}
library(forcats)

ex_theme <- vdj_theme() +
  theme(
    legend.position = "none",
    panel.border    = element_rect(size = 1, color = "grey85", fill = NA),
    axis.line       = element_blank(),
    axis.title      = element_blank(),
    axis.text       = element_blank(),
    axis.ticks      = element_blank()
  )

# PC example
pc_mat <- so_tcr@reductions$pca@cell.embeddings
pcs <- colnames(pc_mat)

pc_df <- pc_mat %>%
  as_tibble(rownames = ".cell_id") %>%
  pivot_longer(cols = c(-.cell_id, -PC_1)) %>%
  mutate(name = fct_relevel(name, pcs))

pc_ex <- pc_df %>%
  filter(name %in% pcs[1:20]) %>%
  ggplot(aes(PC_1, value, color = name)) +
  geom_point(size = 0.1, color = "#56B4E9") +
  facet_wrap(~ name, scales = "free") +
  ex_theme

umap_ex <- so_tcr@meta.data %>%
  ggplot(aes(UMAP_1, UMAP_2)) +
  geom_point(size = 0.5, color = "#009E73") +
  ex_theme
```

