<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BCR analysis • djvdj</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BCR analysis">
<meta property="og:description" content="djvdj">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">djvdj</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bcr.html">BCR analysis</a>
    </li>
    <li>
      <a href="../articles/tcr.html">TCR analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="bcr_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="bcr_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="bcr_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BCR analysis</h1>
            
            <h4 class="date">October 27 2020</h4>
      
      
      <div class="hidden name"><code>bcr.Rmd</code></div>

    </div>

    
    
<p><br></p>
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. In nec molestie risus. Duis vitae eros odio. Proin condimentum odio dolor, at vehicula lorem tempus sit amet. Nunc id metus vehicula, facilisis velit in, tincidunt augue. Donec at semper turpis, nec cursus justo. Suspendisse est lorem, vulputate vitae diam sit amet, luctus volutpat sem. Aenean at odio sed nibh tempor eleifend. Duis ultrices turpis sit amet velit luctus, eu commodo urna lacinia. In porttitor tristique quam ac molestie. In s</p>
<p><img src="bcr_files/figure-html/rna_umap-1.png" width="1200"></p>
<p><br></p>
<div id="import-vdj-data" class="section level3">
<h3 class="hasAnchor">
<a href="#import-vdj-data" class="anchor"></a>Import VDJ data</h3>
<p><code>import_vdj</code> takes the output files from <code>cellranger vdj</code> and adds clonotype information to the meta.data for an existing Seurat object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a “;” (or a character specified by <code>sep</code>). For cells that do not have any VDJ sequencing data, NAs will be added to the meta.data.</p>
<p>If the Seurat object contains data for multiple runs, a vector containing paths to the VDJ data for each sample can be given. If multiple paths are provided, cell prefixes should be included as names for the vector.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># Create vector of paths for cellranger output</span>
<span class="va">paths</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">"data/bcr"</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"[0-9]_AB"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">paths</span>, <span class="st">"_b_Cell$|_bCellEnriched$"</span><span class="op">)</span>
<span class="va">paths</span>        <span class="op">&lt;-</span> <span class="fu">map_chr</span><span class="op">(</span><span class="va">paths</span>, <span class="op">~</span> <span class="fu">here</span><span class="op">(</span><span class="st">"data/bcr"</span>, <span class="va">.x</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Import VDJ data</span>
<span class="va">so_bcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>
  sobj_in        <span class="op">=</span> <span class="va">so_bcr</span>,  <span class="co"># Seurat object</span>
  vdj_dir        <span class="op">=</span> <span class="va">paths</span>,   <span class="co"># Directories containing cellranger output files</span>
  prefix         <span class="op">=</span> <span class="st">""</span>,      <span class="co"># Prefix to add to new meta.data columns</span>
  filter_contigs <span class="op">=</span> <span class="cn">TRUE</span>     <span class="co"># Only include chains with at least one productive contig</span>
<span class="op">)</span>

<span class="co"># Take a look at the meta.data</span>
<span class="va">vdj_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">"clonotype_id"</span>, <span class="st">"cdr3"</span>,
  <span class="st">"chains"</span>, <span class="st">"v_gene"</span>, 
  <span class="st">"j_gene"</span>, <span class="st">"reads"</span>,
  <span class="st">"umis"</span>
<span class="op">)</span>

<span class="va">so_bcr</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 15,438 x 7</span>
<span class="co">#&gt;    clonotype_id     cdr3              chains  v_gene      j_gene  reads   umis  </span>
<span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;             &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt; </span>
<span class="co">#&gt;  1 1_ABA_015_clono… CARVFRDLAVAGTLWF… IGH;IGL IGHV4-61;I… IGHJ5;… 1464;1… 11;26 </span>
<span class="co">#&gt;  2 1_ABA_015_clono… CAKGAREWELRRHSYY… IGH;IGK IGHV3-23;I… IGHJ4;… 2646;8… 24;61 </span>
<span class="co">#&gt;  3 1_ABA_015_clono… CARLGYGDKDYW;CQQ… IGH;IGK IGHV4-59;I… IGHJ4;… 2296;6… 15;38 </span>
<span class="co">#&gt;  4 1_ABA_015_clono… CTSPSKHYYYYYMDVW… IGH;IGL IGHV3-73;I… IGHJ6;… 2996;3… 24;42 </span>
<span class="co">#&gt;  5 1_ABA_015_clono… CARQSGTYDYW;CQAW… IGH;IGL IGHV1-2;IG… IGHJ4;… 1112;5… 11;59 </span>
<span class="co">#&gt;  6 1_ABA_015_clono… CVREGITMIVVSTLSY… IGH;IGL IGHV3-7;IG… IGHJ4;… 36010;… 4584;…</span>
<span class="co">#&gt;  7 1_ABA_015_clono… CARIVVKNYYDSSGYY… IGH;IGK IGHV2-26;I… IGHJ4;… 3542;3… 26;26 </span>
<span class="co">#&gt;  8 1_ABA_015_clono… CAVHGTGPIIMPPGNY… IGH;IGK IGHV1-46;I… IGHJ6;… 27246;… 5345;…</span>
<span class="co">#&gt;  9 1_ABA_015_clono… CARDGGRWYPGAFDIW… IGH;IGK IGHV3-21;I… IGHJ3;… 3706;7… 30;47 </span>
<span class="co">#&gt; 10 1_ABA_015_clono… CARGGATMGPDGAFDI… IGH;IGL IGHV4-61;I… IGHJ3;… 3244;8… 22;89 </span>
<span class="co">#&gt; # … with 15,428 more rows</span></pre></div>
<p><br></p>
</div>
<div id="quality-control" class="section level3">
<h3 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality Control</h3>
<div id="paired-chains" class="section level4">
<h4 class="hasAnchor">
<a href="#paired-chains" class="anchor"></a>Paired Chains</h4>
<p>The V(D)J data imported from Cell Ranger will include clonotypes that do not have paired alpha and beta chains. Using the <code>mutate_vdj</code> function, we can add a new cell label to the meta.data to allow for easy visualization of these cells.</p>
<p>In this example we create a new column that summarizes the unique chains identified for each cell. We can then overlay these cell labels on a UMAP projection.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">so_bcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mutate_vdj.html">mutate_vdj</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_bcr</span>,                                <span class="co"># Seurat object</span>
  uniq_chains <span class="op">=</span> <span class="fu">str_c</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chains</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>  <span class="co"># New column</span>
<span class="op">)</span>

<span class="co"># Take a look at the meta.data</span>
<span class="va">so_bcr</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span>, <span class="va">n_chains</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span>, <span class="va">uniq_chains</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1,095 x 8</span>
<span class="co">#&gt;    clonotype_id  cdr3        chains  v_gene    j_gene   reads  umis  uniq_chains</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;      </span>
<span class="co">#&gt;  1 1_ABA_015_cl… CARDGYGSGI… IGH;IG… IGHV1-18… IGHJ4;I… 4850;… 29;3… IGH_IGK_IGL</span>
<span class="co">#&gt;  2 1_ABA_015_cl… CAREGYYYDS… IGH;IG… IGHV4-61… IGHJ4;I… 1688;… 11;3… IGH_IGL    </span>
<span class="co">#&gt;  3 1_ABA_015_cl… CARGLFGGTQ… IGH;IG… IGHV3-11… IGHJ4;I… 3960;… 28;1… IGH_IGK_IGL</span>
<span class="co">#&gt;  4 1_ABA_015_cl… CATYDSSGPW… IGH;IG… IGHV1-46… IGHJ3;I… 4550;… 35;1… IGH_IGK_IGL</span>
<span class="co">#&gt;  5 1_ABA_015_cl… CAKGDYYDSS… IGH;IG… IGHV3-23… IGHJ6;I… 438;5… 4;4;… IGH_IGK    </span>
<span class="co">#&gt;  6 1_ABA_015_cl… CARDPFSVTM… IGH;IG… IGHV1-69… IGHJ4;I… 3742;… 29;1… IGH_IGK_IGL</span>
<span class="co">#&gt;  7 1_ABA_015_cl… CARRSSAAPR… IGH;IG… IGHV5-51… IGHJ6;I… 3422;… 24;5… IGH_IGK_IGL</span>
<span class="co">#&gt;  8 1_ABA_015_cl… CARVKSWFDP… IGH;IG… IGHV1-2;… IGHJ5;I… 5538;… 48;2… IGH_IGK    </span>
<span class="co">#&gt;  9 1_ABA_015_cl… CAKHKVATMY… IGH;IG… IGHV3-23… IGHJ4;I… 3572;… 25;3… IGH_IGK    </span>
<span class="co">#&gt; 10 1_ABA_015_cl… CTTADFWSGS… IGH;IG… IGHV3-15… IGHJ4;I… 1958;… 15;6… IGH_IGK_IGL</span>
<span class="co">#&gt; # … with 1,085 more rows</span></pre></div>
<p><img src="bcr_files/figure-html/chains_umap-1.png" width="768"></p>
<p><br></p>
<p>Clonotypes that lack paired alpha and beta chains can be removed from the meta.data using <code>filter_vdj</code>. With this function cells that match the provided condition are filtered from the object. Setting <code>filter_cells</code> to <code>FALSE</code> will just remove the V(D)J data without completely removing these cells.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">so_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in      <span class="op">=</span> <span class="va">so_bcr</span>,
  filt         <span class="op">=</span> <span class="st">"IGH"</span> <span class="op">%in%</span> <span class="va">chains</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"IGL"</span>, <span class="st">"IGK"</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">chains</span><span class="op">)</span>,
  filter_cells <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="co"># Take a look at the meta.data</span>
<span class="va">so_filt</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 14,735 x 7</span>
<span class="co">#&gt;    clonotype_id     cdr3              chains  v_gene      j_gene  reads   umis  </span>
<span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;             &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt; </span>
<span class="co">#&gt;  1 1_ABA_015_clono… CARVFRDLAVAGTLWF… IGH;IGL IGHV4-61;I… IGHJ5;… 1464;1… 11;26 </span>
<span class="co">#&gt;  2 1_ABA_015_clono… CAKGAREWELRRHSYY… IGH;IGK IGHV3-23;I… IGHJ4;… 2646;8… 24;61 </span>
<span class="co">#&gt;  3 1_ABA_015_clono… CARLGYGDKDYW;CQQ… IGH;IGK IGHV4-59;I… IGHJ4;… 2296;6… 15;38 </span>
<span class="co">#&gt;  4 1_ABA_015_clono… CTSPSKHYYYYYMDVW… IGH;IGL IGHV3-73;I… IGHJ6;… 2996;3… 24;42 </span>
<span class="co">#&gt;  5 1_ABA_015_clono… CARQSGTYDYW;CQAW… IGH;IGL IGHV1-2;IG… IGHJ4;… 1112;5… 11;59 </span>
<span class="co">#&gt;  6 1_ABA_015_clono… CVREGITMIVVSTLSY… IGH;IGL IGHV3-7;IG… IGHJ4;… 36010;… 4584;…</span>
<span class="co">#&gt;  7 1_ABA_015_clono… CARIVVKNYYDSSGYY… IGH;IGK IGHV2-26;I… IGHJ4;… 3542;3… 26;26 </span>
<span class="co">#&gt;  8 1_ABA_015_clono… CAVHGTGPIIMPPGNY… IGH;IGK IGHV1-46;I… IGHJ6;… 27246;… 5345;…</span>
<span class="co">#&gt;  9 1_ABA_015_clono… CARDGGRWYPGAFDIW… IGH;IGK IGHV3-21;I… IGHJ3;… 3706;7… 30;47 </span>
<span class="co">#&gt; 10 1_ABA_015_clono… CARGGATMGPDGAFDI… IGH;IGL IGHV4-61;I… IGHJ3;… 3244;8… 22;89 </span>
<span class="co">#&gt; # … with 14,725 more rows</span></pre></div>
<p><br></p>
</div>
<div id="read-support" class="section level4">
<h4 class="hasAnchor">
<a href="#read-support" class="anchor"></a>Read Support</h4>
<p>The read support for each chain can be visualized with the <code>plot_reads</code> function. This will create plots summarizing the number of UMIs and total reads that were obtained for each chain.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="../reference/plot_reads.html">plot_reads</a></span><span class="op">(</span>
  sobj_in      <span class="op">=</span> <span class="va">so_bcr</span>,        <span class="co"># Seurat object</span>
  chain_col    <span class="op">=</span> <span class="st">"chains"</span>,      <span class="co"># Column containing chains for each cell</span>
  cluster_col  <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># Column containing labels to group by</span>
  plot_colors  <span class="op">=</span> <span class="va">ito_cols</span>       <span class="co"># Plot colors</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="bcr_files/figure-html/read_support-1.png" width="960"></p>
<p><br></p>
<p>Clonotypes that do not have adequate UMI support can be filtered from the object using <code>filter_vdj</code>. In this example we filter for clonotypes that have chains supported by at least two UMIs.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">so_bcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in      <span class="op">=</span> <span class="va">so_bcr</span>,         <span class="co"># Seurat object</span>
  filt         <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="va">umis</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span>,  <span class="co"># Condition for filtering</span>
  filter_cells <span class="op">=</span> <span class="cn">FALSE</span>           <span class="co"># Should cells be removed from object</span>
<span class="op">)</span>

<span class="co"># Take a look at the meta.data</span>
<span class="va">so_bcr</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 15,260 x 7</span>
<span class="co">#&gt;    clonotype_id     cdr3              chains  v_gene      j_gene  reads   umis  </span>
<span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;             &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt; </span>
<span class="co">#&gt;  1 1_ABA_015_clono… CARVFRDLAVAGTLWF… IGH;IGL IGHV4-61;I… IGHJ5;… 1464;1… 11;26 </span>
<span class="co">#&gt;  2 1_ABA_015_clono… CAKGAREWELRRHSYY… IGH;IGK IGHV3-23;I… IGHJ4;… 2646;8… 24;61 </span>
<span class="co">#&gt;  3 1_ABA_015_clono… CARLGYGDKDYW;CQQ… IGH;IGK IGHV4-59;I… IGHJ4;… 2296;6… 15;38 </span>
<span class="co">#&gt;  4 1_ABA_015_clono… CTSPSKHYYYYYMDVW… IGH;IGL IGHV3-73;I… IGHJ6;… 2996;3… 24;42 </span>
<span class="co">#&gt;  5 1_ABA_015_clono… CARQSGTYDYW;CQAW… IGH;IGL IGHV1-2;IG… IGHJ4;… 1112;5… 11;59 </span>
<span class="co">#&gt;  6 1_ABA_015_clono… CVREGITMIVVSTLSY… IGH;IGL IGHV3-7;IG… IGHJ4;… 36010;… 4584;…</span>
<span class="co">#&gt;  7 1_ABA_015_clono… CARIVVKNYYDSSGYY… IGH;IGK IGHV2-26;I… IGHJ4;… 3542;3… 26;26 </span>
<span class="co">#&gt;  8 1_ABA_015_clono… CAVHGTGPIIMPPGNY… IGH;IGK IGHV1-46;I… IGHJ6;… 27246;… 5345;…</span>
<span class="co">#&gt;  9 1_ABA_015_clono… CARDGGRWYPGAFDIW… IGH;IGK IGHV3-21;I… IGHJ3;… 3706;7… 30;47 </span>
<span class="co">#&gt; 10 1_ABA_015_clono… CARGGATMGPDGAFDI… IGH;IGL IGHV4-61;I… IGHJ3;… 3244;8… 22;89 </span>
<span class="co">#&gt; # … with 15,250 more rows</span></pre></div>
<p><br></p>
</div>
</div>
<div id="clonotype-abundance" class="section level3">
<h3 class="hasAnchor">
<a href="#clonotype-abundance" class="anchor"></a>Clonotype Abundance</h3>
<p>To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the <code>calc_abundance</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">so_bcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_abundance.html">calc_abundance</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_bcr</span>,        <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,        <span class="co"># meta.data column containing clonotype IDs</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># meta.data column containing cell labels</span>
  prefix        <span class="op">=</span> <span class="st">""</span>             <span class="co"># Prefix to add to new meta.data columns</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>For each ‘calculation’ function provided by <code>djvdj</code>, there is a matching <code>plot</code> function that will generate a summary plot. The <code>calc_abundance</code> function will plot clonotypes ranked by abundance.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="../reference/plot_abundance.html">plot_abundance</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_bcr</span>,        <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,        <span class="co"># meta.data column containing clonotype IDs</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># meta.data column containing cell labels</span>
  
  plot_colors <span class="op">=</span> <span class="va">ito_cols</span>,        <span class="co"># Plot colors</span>
  yaxis       <span class="op">=</span> <span class="st">"percent"</span>,       <span class="co"># Units to plot</span>
  label_col   <span class="op">=</span> <span class="st">"cdr3"</span>,          <span class="co"># meta.data column containing labels</span>
  n_labels    <span class="op">=</span> <span class="fl">0</span>,               <span class="co"># Number of top clonotypes to label</span>
  size        <span class="op">=</span> <span class="fl">1</span>                <span class="co"># Additional ggplot options</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="bcr_files/figure-html/abund_plots-1.png" width="576"></p>
<p><br></p>
</div>
<div id="repertoire-diversity" class="section level3">
<h3 class="hasAnchor">
<a href="#repertoire-diversity" class="anchor"></a>Repertoire Diversity</h3>
<p>The function <code>calc_diversity</code> will calculate repertoire diversity on a per-cluster basis. Using the <code>cluster_col</code> argument, any meta.data column containing cell labels can be used for calculations. <code>calc_diversity</code> uses the R package <code>abdiv</code> for performing diversity calculations and any <code>abdiv</code> diversity function can be specified using the <code>method</code> argument.</p>
<p>Possible methods for calculating diversity include:</p>
<pre><code> [1] "berger_parker_d"  "brillouin_d"      "dominance"        "heip_e"          
 [5] "invsimpson"       "kempton_taylor_q" "margalef"         "mcintosh_d"      
 [9] "mcintosh_e"       "menhinick"        "pielou_e"         "richness"        
[13] "shannon"          "simpson"          "simpson_e"        "strong"          </code></pre>
<p><br></p>
<p>In this example we are calculating the Shannon diversity for each sample in the orig.ident meta.data column.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="va">so_bcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_diversity.html">calc_diversity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_bcr</span>,         <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,         <span class="co"># meta.data column containing clonotype ids</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,   <span class="co"># meta.data column containing cell labels</span>
  method        <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">shannon</a></span>  <span class="co"># abdiv method to use</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>The <code>plot_diversity</code> function will create plots summarizing repertoire diversity for each sample. A named list of functions to run can also be passed to plot multiple metrics. Four different diversity metrics are shown in the example below.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co"># Metrics to plot</span>
<span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
  <span class="st">"simpson"</span>     <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/simpson.html">simpson</a></span>,
  <span class="st">"shannon"</span>     <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">shannon</a></span>,
  <span class="st">"margalef"</span>    <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/margalef.html">margalef</a></span>,
  <span class="st">"menhinick"</span>   <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/menhinick.html">menhinick</a></span>,
  <span class="st">"brillouin_d"</span> <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">brillouin_d</a></span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_bcr</span>,        <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,        <span class="co"># meta.data column containing clonotype ids</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># meta.data column containing cell labels</span>
  method        <span class="op">=</span> <span class="va">fns</span>,           <span class="co"># abdiv method to use</span>
  plot_colors   <span class="op">=</span> <span class="va">ito_cols</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="bcr_files/figure-html/div_plots-1.png" width="816"></p>
<p><br></p>
</div>
<div id="repertoire-overlap" class="section level3">
<h3 class="hasAnchor">
<a href="#repertoire-overlap" class="anchor"></a>Repertoire Overlap</h3>
<p>To compare repertoires for different samples or clusters, <code>calc_similarity</code> can calculate a variety of different similarity metrics. The <code>cluster_col</code> should be used to specify the meta.data column containing cell labels for comparison. Like <code>calc_diversity</code>, an <code>abdiv</code> function can be specified with the <code>method</code> argument.</p>
<p>Possible methods for calculating repertoire similarity include:</p>
<pre><code> [1] "binomial_deviance"                  "bray_curtis"                       
 [3] "bray_curtis_balanced"               "bray_curtis_gradient"              
 [5] "canberra"                           "chebyshev"                         
 [7] "chord"                              "clark_coefficient_of_divergence"   
 [9] "correlation_distance"               "cosine_distance"                   
[11] "cy_dissimilarity"                   "euclidean"                         
[13] "geodesic_metric"                    "hamming"                           
[15] "hellinger"                          "horn_morisita"                     
[17] "jaccard"                            "jaccard_nestedness"                
[19] "jaccard_turnover"                   "kulczynski_first"                  
[21] "kulczynski_second"                  "kullback_leibler_divergence"       
[23] "manhattan"                          "mean_character_difference"         
[25] "minkowski"                          "modified_mean_character_difference"
[27] "morisita"                           "rms_distance"                      
[29] "rogers_tanimoto"                    "russel_rao"                        
[31] "ruzicka"                            "ruzicka_balanced"                  
[33] "ruzicka_gradient"                   "sokal_michener"                    
[35] "sokal_sneath"                       "sorenson"                          
[37] "sorenson_nestedness"                "sorenson_turnover"                 
[39] "weighted_kulczynski_second"         "yule_dissimilarity"                </code></pre>
<p><br></p>
<p>By default <code>calc_similarity</code> will add a new meta.data column for each comparison. In this example we are calculating the jaccard dissimilarity index for all combinations of cell labels present in the <code>orig.ident</code> column.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="va">so_bcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_similarity.html">calc_similarity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_bcr</span>,          <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,          <span class="co"># meta.data column containing clonotype ids</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,    <span class="co"># meta.data column containing cell labels</span>
  method        <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>,  <span class="co"># abdiv method to use</span>
  prefix        <span class="op">=</span> <span class="st">"jcrd_"</span>,         <span class="co"># Prefix to add to new meta.data columns </span>
  return_seurat <span class="op">=</span> <span class="cn">TRUE</span>             <span class="co"># Return Seurat object with results added to meta.data</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>A heatmap summarizing the results can be generated using the <code>plot_similarity</code> function. Here we are creating two heatmaps, one to compare the different samples and one to compare cell clusters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="va">heat_theme</span> <span class="op">&lt;-</span> <span class="fu">theme</span><span class="op">(</span>
  legend.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
  legend.text  <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># Sample heatmap</span>
<span class="va">ident_heat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_bcr</span>,                 <span class="co"># Seurat object</span>
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,                 <span class="co"># meta.data column containing clonotype IDs</span>
  cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,           <span class="co"># meta.data column containing cell labels</span>
  method        <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>,         <span class="co"># Method to use</span>
  plot_colors   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey90"</span>, <span class="st">"#009E73"</span><span class="op">)</span>  <span class="co"># Plot colors</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="va">heat_theme</span>

<span class="co"># Cluster heatmap</span>
<span class="va">clust_heat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span>
  sobj_in       <span class="op">=</span> <span class="va">so_bcr</span>,
  clonotype_col <span class="op">=</span> <span class="st">"cdr3"</span>,
  cluster_col   <span class="op">=</span> <span class="st">"seurat_clusters"</span>,
  method        <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>,
  plot_colors   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey90"</span>, <span class="st">"#56B4E9"</span><span class="op">)</span>,  
  size          <span class="op">=</span> <span class="fl">0.2</span>,                    <span class="co"># Additional ggplot options</span>
  color         <span class="op">=</span> <span class="st">"white"</span>                 <span class="co"># Additional ggplot options</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="va">heat_theme</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text.x  <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Combine heatmaps</span>
<span class="fu">plot_grid</span><span class="op">(</span><span class="va">ident_heat</span>, <span class="va">clust_heat</span>, align <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></pre></div>
<p><img src="bcr_files/figure-html/sim_plots-1.png" width="960"></p>
<p><br></p>
</div>
<div id="gene-usage" class="section level3">
<h3 class="hasAnchor">
<a href="#gene-usage" class="anchor"></a>Gene Usage</h3>
<p>The V(D)J data imported from Cell Ranger also includes the specific genes detected for each cell. The function <code>calc_usage</code> can be used to calculate the fraction of cells that express different V(D)J genes. This function will produce a table summarizing the results. To only include results for a certain chain, the <code>chain</code> and <code>chain_col</code> arguments can be used to specify the meta.data column containing the chains detected for each cell. By default the results for all chains will be included.</p>
<p>In this example we are summarizing the usage of different V genes for the TRB chain</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu"><a href="../reference/calc_usage.html">calc_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_bcr</span>,        <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,      <span class="co"># meta.data column containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,  <span class="co"># meta.data column containing cell labels</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,         <span class="co"># Chain to use for filtering genes</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>       <span class="co"># meta.data column containing chains identified for each cell</span>
<span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 5</span>
<span class="co">#&gt;   v_gene orig.ident  n_cells  freq   pct</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;         &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 None   1_ABA_015       655   655   100</span>
<span class="co">#&gt; 2 None   2_ABCtr_015    5110  5110   100</span>
<span class="co">#&gt; 3 None   1_ABA_019      4211  4211   100</span>
<span class="co">#&gt; 4 None   2_ABCtr_019    1646  1646   100</span>
<span class="co">#&gt; 5 None   3_ABA_07       1042  1042   100</span>
<span class="co">#&gt; 6 None   4_ABCtr_07     2596  2596   100</span></pre></div>
<p><br></p>
<p>The function <code>plot_usage</code> can be used to summarize these results. Using the <code>yaxis</code> argument, the percent or absolute count (frequency) can be used for plotting. The genes plotted can also be selected using the <code>plot_genes</code> argument, or the number of top genes (most frequent) to plot can be specified with <code>n_genes</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu"><a href="../reference/plot_usage.html">plot_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_bcr</span>,                <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,              <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,          <span class="co"># meta.data column containing cell labels</span>
  type        <span class="op">=</span> <span class="st">"bar"</span>,                 <span class="co"># Type of plot</span>
  chain       <span class="op">=</span> <span class="st">"IGH"</span>,                 <span class="co"># Chain to use for filtering genes</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>,              <span class="co"># meta.data column containing chains</span>
  
  yaxis       <span class="op">=</span> <span class="st">"percent"</span>,             <span class="co"># Units to plot</span>
  plot_colors <span class="op">=</span> <span class="va">ito_cols</span>,              <span class="co"># Colors to use for heatmap</span>
  plot_genes  <span class="op">=</span> <span class="cn">NULL</span>,                  <span class="co"># A list of genes to plot</span>
  n_genes     <span class="op">=</span> <span class="fl">10</span>,                    <span class="co"># The number of top genes to plot</span>
  
  size        <span class="op">=</span> <span class="fl">0.2</span>,                   <span class="co"># Additional ggplot options</span>
  color       <span class="op">=</span> <span class="st">"white"</span>                <span class="co"># Additional ggplot options</span>
<span class="op">)</span></pre></div>
<p><img src="bcr_files/figure-html/usage_plots_1-1.png" width="1056"></p>
<p><br></p>
<p>By passing multiple columns to <code>gene_cols</code>, the frequency that different genes are used together can also be summarized.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu"><a href="../reference/calc_usage.html">calc_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_bcr</span>,                 <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,  <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,           <span class="co"># meta.data column containing cell labels</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,                  <span class="co"># Chain to use for filtering genes</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>                <span class="co"># meta.data column containing chains</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>When multiple gene columns are passed to <code>plot_usage</code>, a list of plots will be returned, one for each cell label in the <code>cluster_col</code> column.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="va">ggs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_usage.html">plot_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_bcr</span>,                 <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,  <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,           <span class="co"># meta.data column containing cell labels</span>
  chain       <span class="op">=</span> <span class="st">"IGH"</span>,                  <span class="co"># Chain to use for filtering genes</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>,               <span class="co"># meta.data column containing chains identified</span>
  plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey90"</span>, <span class="st">"#6A51A3"</span><span class="op">)</span>  <span class="co"># Colors to use for heatmap</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">imap</span><span class="op">(</span><span class="op">~</span> <span class="va">.x</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="va">.y</span><span class="op">)</span><span class="op">)</span>

<span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">ggs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></pre></div>
<p><img src="bcr_files/figure-html/usage_plots_2-1.png" width="1152"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ryan Sheridan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
