<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>TCR analysis • djvdj</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="TCR analysis">
<meta property="og:description" content="djvdj">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">djvdj</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tcr.html">TCR analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tcr_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="tcr_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="tcr_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>TCR analysis</h1>
            
            <h4 class="date">November 11 2020</h4>
      
      
      <div class="hidden name"><code>tcr.Rmd</code></div>

    </div>

    
    
<p><br></p>
<p>For this vignette we are using single-cell TCR data provided by the Rincon lab. Thymocytes were collected for wild type and mutant mice and sorted for two developmental timepoints, DN3 and DN4. UMAP projections are shown below with cells colored by sample and cell cluster (based on gene expression).</p>
<p><img src="tcr_files/figure-html/rna_umap-1.png" width="1200"></p>
<p><br></p>
<div id="import-vdj-data" class="section level2">
<h2 class="hasAnchor">
<a href="#import-vdj-data" class="anchor"></a>Import VDJ data</h2>
<p><code><a href="../reference/import_vdj.html">import_vdj()</a></code> takes the output files from <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/vdj#header">Cell Ranger</a> and adds clonotype information to the meta.data for an existing <a href="https://satijalab.org/seurat/">Seurat</a> object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a ‘;’ (or a character specified by <code>sep</code>). For cells that do not have any VDJ sequencing data, <code>NA</code>s will be added to the meta.data.</p>
<p>If the Seurat object contains data for multiple runs, a vector containing paths to the VDJ data for each sample can be given. If multiple paths are provided, cell prefixes should be included as names for the vector.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># Create vector of paths pointing to cellranger output</span>
<span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  KI_DN3_GE <span class="op">=</span> <span class="st">"../data/tcr/KI_DN3_TCR"</span>,
  KI_DN4_GE <span class="op">=</span> <span class="st">"../data/tcr/KI_DN4_TCR"</span>,
  WT_DN3_GE <span class="op">=</span> <span class="st">"../data/tcr/WT_DN3_TCR"</span>,
  WT_DN4_GE <span class="op">=</span> <span class="st">"../data/tcr/WT_DN4_TCR"</span>
<span class="op">)</span>

<span class="va">so_tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>
  sobj_in        <span class="op">=</span> <span class="va">so_tcr</span>,                  <span class="co"># Seurat object</span>
  vdj_dir        <span class="op">=</span> <span class="va">paths</span>,                   <span class="co"># cellranger directories</span>
  prefix         <span class="op">=</span> <span class="st">""</span>,                      <span class="co"># Prefix to add to new meta.data columns</span>
  filter_contigs <span class="op">=</span> <span class="cn">TRUE</span>                     <span class="co"># Only include productive chains</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>Take a look at the meta.data to see the V(D)J data added to the object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">vdj_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">"clonotype_id"</span>, <span class="st">"cdr3"</span>,
  <span class="st">"chains"</span>, <span class="st">"v_gene"</span>, 
  <span class="st">"j_gene"</span>, <span class="st">"reads"</span>,
  <span class="st">"umis"</span>
<span class="op">)</span>

<span class="va">so_tcr</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3,850 x 7</span>
<span class="co">#&gt;    clonotype_id     cdr3             chains  v_gene      j_gene    reads   umis </span>
<span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;</span>
<span class="co">#&gt;  1 WT_DN3_GE_clono… CAVQGANTEVFF     TRB     TRBV17      TRBJ1-1   42590   72   </span>
<span class="co">#&gt;  2 WT_DN3_GE_clono… CASSHPGQNSGNTLYF TRB     TRBV5       TRBJ1-3   12670   17   </span>
<span class="co">#&gt;  3 WT_DN3_GE_clono… CASSHWGETLYF     TRB     TRBV5       TRBJ2-3   31772   46   </span>
<span class="co">#&gt;  4 WT_DN3_GE_clono… CGARAQGLYNSPLYF  TRB     TRBV20      TRBJ1-6   7124    15   </span>
<span class="co">#&gt;  5 WT_DN3_GE_clono… CTAPAGGQNTEVFF   TRB     TRBV1       TRBJ1-1   4744    8    </span>
<span class="co">#&gt;  6 WT_DN3_GE_clono… CASSQDLDWGGEQFF  TRB     TRBV5       TRBJ2-1   34966   52   </span>
<span class="co">#&gt;  7 WT_DN3_GE_clono… CASRTGGCYEQYF;C… TRB;TRB TRBV13-1;T… TRBJ2-7;… 4166;6… 7;171</span>
<span class="co">#&gt;  8 WT_DN3_GE_clono… CASSPGTENTLYF    TRB     TRBV12-2    TRBJ2-4   9180    18   </span>
<span class="co">#&gt;  9 WT_DN3_GE_clono… CASSLKGARSDYTF   TRB     TRBV12-1    TRBJ1-2   19742   29   </span>
<span class="co">#&gt; 10 WT_DN3_GE_clono… CASRLTGRDSDYTF;… TRB;TRB TRBV15;TRB… TRBJ1-2;… 28860;… 48;22</span>
<span class="co">#&gt; # … with 3,840 more rows</span></pre></div>
<p><br></p>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality Control</h2>
<div id="read-support" class="section level3">
<h3 class="hasAnchor">
<a href="#read-support" class="anchor"></a>Read Support</h3>
<p>The read support for each chain can be visualized with the <code><a href="../reference/plot_reads.html">plot_reads()</a></code> function. This will create plots summarizing the number of UMIs and total reads that were obtained for each chain.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="../reference/plot_reads.html">plot_reads</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>,                   <span class="co"># Column containing chains for each cell</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># Column containing labels to group by</span>
  plot_colors <span class="op">=</span> <span class="va">vdj_colors</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/read_support-1.png" width="624"></p>
<p><br></p>
<p>The djvdj package provides several functions that can be used to manipulate the object meta.data, which is useful for identifying cells that should be removed. This includes a function to modify meta.data columns (<code><a href="../reference/mutate_vdj.html">mutate_vdj()</a></code>) and a function to filter cells from the object (<code><a href="../reference/filter_vdj.html">filter_vdj()</a></code>). Since cells can have multiple values present in each V(D)J column (separated by a semicolon), when using these functions each string of V(D)J values for a cell should be thought of as a vector.</p>
<p>For example using <code><a href="../reference/filter_vdj.html">filter_vdj()</a></code> we can remove V(D)J data for cells that have minimal UMI support. To do this we use the expression <code class="sourceCode r"><span class="kw">all</span>(umis <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)</code>. This means that all semicolon-separated values present in the <code>umis</code> column must be greater than 2. Setting <code class="sourceCode r">filter_cells =<span class="st"> </span><span class="ot">FALSE</span></code> means the cell will remain in the object, but the V(D)J data will be removed.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">so_tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in      <span class="op">=</span> <span class="va">so_tcr</span>,                    <span class="co"># Seurat object</span>
  filt         <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="va">umis</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span>,             <span class="co"># Condition for filtering</span>
  filter_cells <span class="op">=</span> <span class="cn">FALSE</span>                      <span class="co"># Should cells be removed from object</span>
<span class="op">)</span>

<span class="co"># Take a look at the meta.data</span>
<span class="va">so_tcr</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3,725 x 7</span>
<span class="co">#&gt;    clonotype_id     cdr3             chains  v_gene      j_gene    reads   umis </span>
<span class="co">#&gt;    &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;</span>
<span class="co">#&gt;  1 WT_DN3_GE_clono… CAVQGANTEVFF     TRB     TRBV17      TRBJ1-1   42590   72   </span>
<span class="co">#&gt;  2 WT_DN3_GE_clono… CASSHPGQNSGNTLYF TRB     TRBV5       TRBJ1-3   12670   17   </span>
<span class="co">#&gt;  3 WT_DN3_GE_clono… CASSHWGETLYF     TRB     TRBV5       TRBJ2-3   31772   46   </span>
<span class="co">#&gt;  4 WT_DN3_GE_clono… CGARAQGLYNSPLYF  TRB     TRBV20      TRBJ1-6   7124    15   </span>
<span class="co">#&gt;  5 WT_DN3_GE_clono… CTAPAGGQNTEVFF   TRB     TRBV1       TRBJ1-1   4744    8    </span>
<span class="co">#&gt;  6 WT_DN3_GE_clono… CASSQDLDWGGEQFF  TRB     TRBV5       TRBJ2-1   34966   52   </span>
<span class="co">#&gt;  7 WT_DN3_GE_clono… CASRTGGCYEQYF;C… TRB;TRB TRBV13-1;T… TRBJ2-7;… 4166;6… 7;171</span>
<span class="co">#&gt;  8 WT_DN3_GE_clono… CASSPGTENTLYF    TRB     TRBV12-2    TRBJ2-4   9180    18   </span>
<span class="co">#&gt;  9 WT_DN3_GE_clono… CASSLKGARSDYTF   TRB     TRBV12-1    TRBJ1-2   19742   29   </span>
<span class="co">#&gt; 10 WT_DN3_GE_clono… CASRLTGRDSDYTF;… TRB;TRB TRBV15;TRB… TRBJ1-2;… 28860;… 48;22</span>
<span class="co">#&gt; # … with 3,715 more rows</span></pre></div>
<p>To further illustrate how <code><a href="../reference/filter_vdj.html">filter_vdj()</a></code> works we can look at row 7 shown above, which has two beta chains with 7 UMIs for the first and 171 UMIs for the second. When filtering this row we are evaluating whether the following condition is <code>TRUE</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">7</span>, <span class="fl">171</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="va">vec</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></pre></div>
<p><br></p>
</div>
<div id="paired-chains" class="section level3">
<h3 class="hasAnchor">
<a href="#paired-chains" class="anchor"></a>Paired Chains</h3>
<p>The V(D)J data imported from Cell Ranger will include clonotypes that do not have paired alpha and beta chains. Using the <code><a href="../reference/mutate_vdj.html">mutate_vdj()</a></code> function, we can add a new cell label to the meta.data to allow for easy visualization of these cells.</p>
<p>In this example we create a new column (<code>uniq_chains</code>) that summarizes the unique chains identified for each cell. We can then overlay these cell labels on a UMAP projection.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">so_tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mutate_vdj.html">mutate_vdj</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                                <span class="co"># Seurat object</span>
  uniq_chains <span class="op">=</span> <span class="fu">str_c</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">chains</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>  <span class="co"># New column</span>
<span class="op">)</span>

<span class="co"># Take a look at the meta.data</span>
<span class="va">so_tcr</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span>
  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clonotype_id</span><span class="op">)</span>, <span class="va">n_chains</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">vdj_cols</span><span class="op">)</span>, <span class="va">uniq_chains</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 83 x 8</span>
<span class="co">#&gt;    clonotype_id  cdr3       chains  v_gene    j_gene    reads  umis  uniq_chains</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;      &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;      </span>
<span class="co">#&gt;  1 WT_DN4_GE_cl… CALVMNYNQ… TRA;TR… TRAV13-1… TRAJ23;T… 16886… 13;1… TRA_TRB    </span>
<span class="co">#&gt;  2 WT_DN4_GE_cl… CAAYSGGSN… TRA;TR… TRAV14-3… TRAJ53;T… 12106… 14;8… TRA_TRB    </span>
<span class="co">#&gt;  3 WT_DN4_GE_cl… CVVVDLPGT… TRA;TR… TRAV11;T… TRAJ28;T… 6974;… 7;11… TRA_TRB    </span>
<span class="co">#&gt;  4 WT_DN4_GE_cl… CAARRGSNY… TRA;TR… TRAV14D-… TRAJ33;T… 20144… 13;3… TRA_TRB    </span>
<span class="co">#&gt;  5 WT_DN4_GE_cl… CVAHNNAGA… TRA;TR… TRAV11D;… TRAJ39;T… 7230;… 5;10… TRA_TRB    </span>
<span class="co">#&gt;  6 WT_DN4_GE_cl… CAASTSGSW… TRA;TR… TRAV7D-2… TRAJ22;T… 18152… 11;8… TRA_TRB    </span>
<span class="co">#&gt;  7 WT_DN4_GE_cl… CALLASSSF… TRA;TR… TRAV12-1… TRAJ50;T… 12942… 19;5… TRA_TRB    </span>
<span class="co">#&gt;  8 WT_DN4_GE_cl… CAPGTGGYK… TRA;TR… TRAV13-2… TRAJ12;T… 11882… 13;4… TRA_TRB    </span>
<span class="co">#&gt;  9 WT_DN4_GE_cl… CAASALRDS… TRA;TR… TRAV14N-… TRAJ13;T… 13372… 8;15… TRA_TRB    </span>
<span class="co">#&gt; 10 WT_DN4_GE_cl… CAPDTNAYK… TRA;TR… TRAV13D-… TRAJ30;T… 11016… 7;13… TRA_TRB    </span>
<span class="co">#&gt; # … with 73 more rows</span></pre></div>
<p>For row 1 shown above, this is equivalent to:</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TRA"</span>, <span class="st">"TRB"</span>, <span class="st">"TRB"</span><span class="op">)</span>
<span class="va">vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">vec</span><span class="op">)</span>

<span class="fu">str_c</span><span class="op">(</span><span class="va">vec</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
<span class="co">#&gt; [1] "TRA_TRB"</span></pre></div>
<p><br></p>
<p>We can then overlay the new labels present in the <code>uniq_chains</code> column on a UMAP projection. djvdj also provides a simple wrapper function (<code><a href="../reference/plot_cell_count.html">plot_cell_count()</a></code>) that will summarize the fraction of cells that belong to each group.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">chain_cols</span> <span class="op">&lt;-</span> <span class="va">vdj_colors</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">chain_cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">so_tcr</span><span class="op">$</span><span class="va">uniq_chains</span><span class="op">)</span>

<span class="co"># Chains UMAP</span>
<span class="va">chain_umap</span> <span class="op">&lt;-</span> <span class="va">so_tcr</span> <span class="op">%&gt;%</span>
  <span class="fu">DimPlot</span><span class="op">(</span>
    group.by <span class="op">=</span> <span class="st">"uniq_chains"</span>,
    cols     <span class="op">=</span> <span class="va">chain_cols</span>,
    na.value <span class="op">=</span> <span class="st">"grey90"</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_nothing</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span>
  
<span class="co"># Chains bar graphs</span>
<span class="va">chain_bars</span> <span class="op">&lt;-</span> <span class="va">so_tcr</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/plot_cell_count.html">plot_cell_count</a></span><span class="op">(</span>
    x           <span class="op">=</span> <span class="st">"orig.ident"</span>,             <span class="co"># Cell labels to plot</span>
    fill_col    <span class="op">=</span> <span class="st">"uniq_chains"</span>,            <span class="co"># Cell labels to color each bar</span>
    plot_colors <span class="op">=</span> <span class="va">chain_cols</span>,
    yaxis       <span class="op">=</span> <span class="st">"counts"</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>
    plot.margin     <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">120</span>, <span class="fl">10</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>, <span class="st">"pt"</span><span class="op">)</span>,
    legend.position <span class="op">=</span> <span class="st">"none"</span>
  <span class="op">)</span>

<span class="co"># Final figure</span>
<span class="fu">plot_grid</span><span class="op">(</span>
  <span class="va">chain_umap</span>, <span class="va">chain_bars</span>,
  rel_widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.55</span><span class="op">)</span>,
  nrow       <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/chains_umap-1.png" width="624"></p>
<p><br></p>
<p>Clonotypes that lack paired alpha and beta chains can be filtered using <code><a href="../reference/filter_vdj.html">filter_vdj()</a></code>. However, for this dataset we expect a large fraction of cells to have unpaired chains, so we can skip this step.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="va">so_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span>
  sobj_in      <span class="op">=</span> <span class="va">so_tcr</span>,                            <span class="co"># Seurat object</span>
  filt         <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TRA"</span>, <span class="st">"TRB"</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">chains</span><span class="op">)</span>,  <span class="co"># Condition for filtering</span>
  filter_cells <span class="op">=</span> <span class="cn">FALSE</span>                              <span class="co"># Should cells be removed</span>
<span class="op">)</span></pre></div>
<p><br></p>
</div>
</div>
<div id="clonotype-abundance" class="section level2">
<h2 class="hasAnchor">
<a href="#clonotype-abundance" class="anchor"></a>Clonotype Abundance</h2>
<p>To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the <code><a href="../reference/calc_abundance.html">calc_abundance()</a></code> function. These calculations can be performed on a per-cluster or per-sample basis by also providing a meta.data column containing cell labels.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">so_tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_abundance.html">calc_abundance</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>                <span class="co"># meta.data column containing cell labels</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>For each ‘calc’ function provided by djvdj, there is a matching ‘plot’ function that will generate a summary plot. The <code><a href="../reference/plot_abundance.html">plot_abundance()</a></code> function will plot clonotypes ranked by abundance.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="fu"><a href="../reference/plot_abundance.html">plot_abundance</a></span><span class="op">(</span>
  sobj_in      <span class="op">=</span> <span class="va">so_tcr</span>,                    <span class="co"># Seurat object</span>
  cluster_col  <span class="op">=</span> <span class="st">"orig.ident"</span>,              <span class="co"># meta.data column containing cell labels</span>
  label_col    <span class="op">=</span> <span class="st">"cdr3"</span>,                    <span class="co"># meta.data column containing labels</span>
  n_clonotypes <span class="op">=</span> <span class="fl">12</span>,                        <span class="co"># Number of top clonotypes</span>
  plot_colors  <span class="op">=</span> <span class="va">vdj_colors</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>
    legend.position <span class="op">=</span> <span class="st">"none"</span>,
    axis.text.x     <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
  <span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/abund_plots-1.png" width="960"></p>
<p><br></p>
</div>
<div id="repertoire-diversity" class="section level2">
<h2 class="hasAnchor">
<a href="#repertoire-diversity" class="anchor"></a>Repertoire Diversity</h2>
<p>The function <code><a href="../reference/calc_diversity.html">calc_diversity()</a></code> will calculate repertoire diversity on number of cells that share each clonotype. Using the <code>cluster_col</code> argument, any meta.data column containing cell labels can be used for calculations. <code><a href="../reference/calc_diversity.html">calc_diversity()</a></code> uses the R package <a href="https://github.com/kylebittinger/abdiv">abdiv</a> for performing diversity calculations and any <a href="https://github.com/kylebittinger/abdiv">abdiv</a> diversity function can be specified using the <code>method</code> argument.</p>
<p>Possible methods for calculating diversity include:</p>
<pre><code> [1] "berger_parker_d"  "brillouin_d"      "dominance"        "heip_e"          
 [5] "invsimpson"       "kempton_taylor_q" "margalef"         "mcintosh_d"      
 [9] "mcintosh_e"       "menhinick"        "pielou_e"         "richness"        
[13] "shannon"          "simpson"          "simpson_e"        "strong"          </code></pre>
<p><br></p>
<p>In this example we are calculating the Shannon diversity for each sample in the orig.ident meta.data column.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="va">so_tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_diversity.html">calc_diversity</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">shannon</a></span>              <span class="co"># abdiv method to use</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>The <code><a href="../reference/plot_diversity.html">plot_diversity()</a></code> function will create plots summarizing repertoire diversity for each sample. A named list of functions to run can also be passed to plot multiple metrics. Four different diversity metrics are shown in the example below.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="co"># Metrics to plot</span>
<span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
  <span class="st">"simpson"</span>     <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/simpson.html">simpson</a></span>,
  <span class="st">"shannon"</span>     <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">shannon</a></span>,
  <span class="st">"margalef"</span>    <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/margalef.html">margalef</a></span>,
  <span class="st">"menhinick"</span>   <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/menhinick.html">menhinick</a></span>,
  <span class="st">"brillouin_d"</span> <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html">brillouin_d</a></span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  method      <span class="op">=</span> <span class="va">fns</span>,                        <span class="co"># abdiv method to use</span>
  plot_colors <span class="op">=</span> <span class="va">vdj_colors</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/div_plots-1.png" width="720"></p>
<p><br></p>
</div>
<div id="repertoire-overlap" class="section level2">
<h2 class="hasAnchor">
<a href="#repertoire-overlap" class="anchor"></a>Repertoire Overlap</h2>
<p>To compare repertoires for different samples or clusters, <code><a href="../reference/calc_similarity.html">calc_similarity()</a></code> can calculate a variety of different similarity metrics. The <code>cluster_col</code> should be used to specify the meta.data column containing cell labels for comparison. Like <code><a href="../reference/calc_diversity.html">calc_diversity()</a></code>, an <a href="https://github.com/kylebittinger/abdiv">abdiv</a> function can be specified with the <code>method</code> argument.</p>
<p>Possible methods for calculating repertoire similarity include:</p>
<pre><code> [1] "binomial_deviance"                  "bray_curtis"                       
 [3] "bray_curtis_balanced"               "bray_curtis_gradient"              
 [5] "canberra"                           "chebyshev"                         
 [7] "chord"                              "clark_coefficient_of_divergence"   
 [9] "correlation_distance"               "cosine_distance"                   
[11] "cy_dissimilarity"                   "euclidean"                         
[13] "geodesic_metric"                    "hamming"                           
[15] "hellinger"                          "horn_morisita"                     
[17] "jaccard"                            "jaccard_nestedness"                
[19] "jaccard_turnover"                   "kulczynski_first"                  
[21] "kulczynski_second"                  "kullback_leibler_divergence"       
[23] "manhattan"                          "mean_character_difference"         
[25] "minkowski"                          "modified_mean_character_difference"
[27] "morisita"                           "rms_distance"                      
[29] "rogers_tanimoto"                    "russel_rao"                        
[31] "ruzicka"                            "ruzicka_balanced"                  
[33] "ruzicka_gradient"                   "sokal_michener"                    
[35] "sokal_sneath"                       "sorenson"                          
[37] "sorenson_nestedness"                "sorenson_turnover"                 
[39] "weighted_kulczynski_second"         "yule_dissimilarity"                </code></pre>
<p><br></p>
<p>By default <code><a href="../reference/calc_similarity.html">calc_similarity()</a></code> will add a new meta.data column for each comparison. In this example we are calculating the jaccard dissimilarity index for all combinations of cell labels present in the <code>orig.ident</code> column.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="va">so_tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_similarity.html">calc_similarity</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>              <span class="co"># abdiv method to use</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>A heatmap summarizing the results can be generated using the <code><a href="../reference/plot_similarity.html">plot_similarity()</a></code> function. Here we are creating two heatmaps, one to compare the different samples and one to compare cell clusters.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="co"># Sample heatmap</span>
<span class="va">ident_heat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>,             <span class="co"># Method to use</span>
  plot_colors <span class="op">=</span> <span class="st">"#009E73"</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Cluster heatmap</span>
<span class="va">clust_heat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>,
  method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html">jaccard</a></span>,
  plot_colors <span class="op">=</span> <span class="st">"#56B4E9"</span>,
  size        <span class="op">=</span> <span class="fl">0.2</span>,                        <span class="co"># Additional ggplot options</span>
  color       <span class="op">=</span> <span class="st">"white"</span>                     <span class="co"># Additional ggplot options</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>
    legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,
    axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  <span class="op">)</span>

<span class="co"># Combine heatmaps</span>
<span class="fu">plot_grid</span><span class="op">(</span><span class="va">ident_heat</span>, <span class="va">clust_heat</span>, align <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/sim_plots-1.png" width="960"></p>
<p><br></p>
</div>
<div id="gene-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-usage" class="anchor"></a>Gene Usage</h2>
<p>The V(D)J data imported from Cell Ranger also includes the specific genes detected for each cell. The function <code><a href="../reference/calc_usage.html">calc_usage()</a></code> can be used to calculate the fraction of cells that express different V(D)J genes. This function will produce a table summarizing the results. To only include results for a certain chain, the <code>chain</code> and <code>chain_col</code> arguments can be used to specify the meta.data column containing the chains detected for each cell. By default the results for all chains will be included.</p>
<p>In this example we are summarizing the usage of different V genes for the TRB chain</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="fu"><a href="../reference/calc_usage.html">calc_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,                   <span class="co"># meta.data column containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,                      <span class="co"># Chain to use for filtering genes</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>                    <span class="co"># meta.data column containing chains</span>
<span class="op">)</span>
<span class="co">#&gt; # A tibble: 92 x 5</span>
<span class="co">#&gt;    v_gene orig.ident n_cells  freq   pct</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;        &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 None   WT_DN3         592     0 0    </span>
<span class="co">#&gt;  2 None   WT_DN4         866    11 1.27 </span>
<span class="co">#&gt;  3 None   KI_DN3         719     0 0    </span>
<span class="co">#&gt;  4 None   KI_DN4        1548     6 0.388</span>
<span class="co">#&gt;  5 TRBV1  WT_DN3         592    23 3.89 </span>
<span class="co">#&gt;  6 TRBV1  WT_DN4         866    37 4.27 </span>
<span class="co">#&gt;  7 TRBV1  KI_DN3         719    34 4.73 </span>
<span class="co">#&gt;  8 TRBV1  KI_DN4        1548    64 4.13 </span>
<span class="co">#&gt;  9 TRBV10 WT_DN3         592    31 5.24 </span>
<span class="co">#&gt; 10 TRBV10 WT_DN4         866    57 6.58 </span>
<span class="co">#&gt; # … with 82 more rows</span></pre></div>
<p><br></p>
<p>The function <code><a href="../reference/plot_usage.html">plot_usage()</a></code> can be used to summarize these results. Using the <code>yaxis</code> argument, the percent or absolute count (frequency) can be used for plotting. The genes plotted can also be selected using the <code>plot_genes</code> argument, or the number of top genes (most frequent) to plot can be specified with <code>n_genes</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="fu"><a href="../reference/plot_usage.html">plot_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,                   <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  type        <span class="op">=</span> <span class="st">"bar"</span>,                      <span class="co"># Type of plot</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,                      <span class="co"># Chain to use for filtering genes</span>
  
  plot_genes  <span class="op">=</span> <span class="cn">NULL</span>,                       <span class="co"># A list of genes to plot</span>
  n_genes     <span class="op">=</span> <span class="cn">NULL</span>,                       <span class="co"># The number of top genes to plot</span>
  plot_colors <span class="op">=</span> <span class="va">vdj_colors</span>
<span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/gene_usage_1-1.png" width="1056"></p>
<p><br></p>
<p>By passing multiple columns to <code>gene_cols</code>, the frequency that different genes are used together can also be summarized.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="fu"><a href="../reference/calc_usage.html">calc_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,      <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>                       <span class="co"># Chain to use for filtering genes</span>
<span class="op">)</span></pre></div>
<p><br></p>
<p>When multiple gene columns are passed to <code><a href="../reference/plot_usage.html">plot_usage()</a></code>, a list of plots will be returned, one for each cell label in the <code>cluster_col</code> column.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="va">ggs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_usage.html">plot_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,                     <span class="co"># Seurat object</span>
  gene_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,      <span class="co"># meta.data column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,               <span class="co"># meta.data column containing cell labels</span>
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,                      <span class="co"># Chain to use for filtering genes</span>
  plot_colors <span class="op">=</span> <span class="st">"#6A51A3"</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">imap</span><span class="op">(</span><span class="op">~</span> <span class="va">.x</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="va">.y</span><span class="op">)</span><span class="op">)</span>

<span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">ggs</span><span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/gene_usage_2-1.png" width="1152"></p>
<p><br></p>
</div>
<div id="cdr3-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#cdr3-usage" class="anchor"></a>CDR3 Usage</h2>
<p>The <code><a href="../reference/calc_usage.html">calc_usage()</a></code> and <code><a href="../reference/plot_usage.html">plot_usage()</a></code> functions can also be used to calculate the fraction of cells that have a certain CDR3 sequence. To do this, the meta.data column containing the CDR3 sequences for each cell can be passed to the <code>gene_cols</code> argument. The plot can be divided into facets based on sample using simple ggplot2 commands.</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="fu"><a href="../reference/plot_usage.html">plot_usage</a></span><span class="op">(</span>
  sobj_in     <span class="op">=</span> <span class="va">so_tcr</span>,
  gene_cols   <span class="op">=</span> <span class="st">"cdr3"</span>,                     <span class="co"># Column containing CDR3 sequences</span>
  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,
  
  type        <span class="op">=</span> <span class="st">"bar"</span>,
  chain       <span class="op">=</span> <span class="st">"TRB"</span>,
  plot_colors <span class="op">=</span> <span class="va">vdj_colors</span>,
  n_genes     <span class="op">=</span> <span class="fl">12</span>                          <span class="co"># Number of CDR3 sequences to plot</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">orig.ident</span>, nrow <span class="op">=</span> <span class="fl">1</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></pre></div>
<p><img src="tcr_files/figure-html/cdr3_usage-1.png" width="960"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ryan Sheridan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
